{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "import cobra\n",
    "import cobra.test\n",
    "# import mackinac\n",
    "import numpy as np\n",
    "import csv\n",
    "import glob\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import time\n",
    "import sys\n",
    "from copy import deepcopy\n",
    "from collections import defaultdict\n",
    "# from cobra.flux_analysis import gapfill\n",
    "from gapfill2 import *\n",
    "from cobra.flux_analysis import pfba\n",
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "from cobra.flux_analysis import sample\n",
    "from cobra.core.solution import get_solution\n",
    "\n",
    "\n",
    "# Set default logger to python logger to avoid warnings given when adding reactions and/or metaboites \n",
    "# because \"cobra.core.model\" doesn't innately have a logger.\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger('logger')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_media(model, media, universal, verbose=False):\n",
    "\n",
    "    # Find and close all exchange reactions in the model\n",
    "    model_rxns = [rxn.id for rxn in model.reactions]\n",
    "    for rxn in model_rxns:\n",
    "        if rxn.startswith('EX_') and rxn.endswith('_e'):\n",
    "            model.reactions.get_by_id(rxn).lower_bound = 0.0\n",
    "\n",
    "    # Check for existence of exchange reactions for the media metabolites in the model\n",
    "    for metabolite in media:\n",
    "        met = metabolite[1]+'_e'\n",
    "        if 'EX_'+met in model_rxns:\n",
    "            model.reactions.get_by_id('EX_'+met).lower_bound = -1000.\n",
    "        else:\n",
    "            # Create exchange reaction and add to model\n",
    "            if verbose:\n",
    "                print(\"added exchange rxn for \" + met)\n",
    "            new_exchange = cobra.Reaction('EX_'+met)\n",
    "            new_exchange.name = met + ' exchange'\n",
    "            met_obj = universal.metabolites.get_by_id(met)\n",
    "            new_exchange.add_metabolites({met_obj:-1})\n",
    "            new_exchange.lower_bound = -1000.\n",
    "            new_exchange.upper_bound = 1000.\n",
    "            model.add_reaction(new_exchange)\n",
    "            model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basal Synthetic Media\n",
    "bsm = [\n",
    "    ['H+','cpd00067'],\n",
    "    ['H2O','cpd00001'],\n",
    "    ['CO2','cpd00011'],\n",
    "    ['O2','cpd00007'],\n",
    "    ['N2','cpd00528'], \n",
    "#     ['H2','cpd11640'], # Only with no O2\n",
    "    \n",
    "    ['K+','cpd00205'],\n",
    "    ['Na+','cpd00971'],\n",
    "    ['Mg','cpd00254'],\n",
    "    ['Mn2+','cpd00030'],\n",
    "    ['Fe2+','cpd10515'], # Iron ion in heme\n",
    "    ['Ca2+','cpd00063'], # Calcium pantothenate;cpd19112\n",
    "    \n",
    "    ['Vitamin B12r','cpd00423'], # C62H91CoN13O14P : cobalamin;cpd03424;cpd00730 : not present in any exchange reactions\n",
    "    ['Cobinamide','cpd03422'], #EXs : related to cobalamin (B12) Added to ensure cells have access to B12\n",
    "    ['BIOT','cpd00104'], # C10H15N2O3S : biotin B7\n",
    "    ['PAN','cpd00644'], # C9H16NO5 : Pantothenate B5\n",
    "    ['Folate','cpd00393'], # C19H17N7O6 : B9\n",
    "    ['Niacin','cpd00218'], # C6H4NO2 : B3\n",
    "    ['Pyridoxal','cpd00215'], # C8H9NO3 : B6\n",
    "    ['Riboflavin','cpd00220'], # C17H19N4O6 : B2\n",
    "    ['thiamin','cpd00305'], # C12H17N4OS : B1\n",
    "    \n",
    "#     ['Phosphate','cpd00009'], # HO4P : In M9 Defaults\n",
    "    \n",
    "    ['Thioglycolate','cpd01415'], # C2H3O2S : not present in any exchange reactions\n",
    "#     ['Sulfate','cpd00048'], # O4S : In M9 Defaults\n",
    "    \n",
    "    ['Acetate','cpd00029'], # C2H3O2 : not present in any exchange reactions\n",
    "    ['Citrate','cpd00137'], # C6H5O7 : Consider removing. \n",
    "#     ['Polysorbate 60','cpd24450'], # C35H68O10 : Almost tween 80 : not present in any reactions\n",
    "#     ['Ethyl acetate','cpd00633'], # C4H8O2 : not present in any exchange reactions, only present in one reaction at all\n",
    "    \n",
    "    ['ABEE','cpd00443'] # C7H6NO2 : aminobenzoate : not present in any exchange reactions\n",
    "]\n",
    "\n",
    "# Potentially add to BSM (from M9 media)\n",
    "M9_ions = [\n",
    "    ['Cl-','cpd00099'],\n",
    "    ['Co2+','cpd00149'],\n",
    "    ['Cu2+','cpd00058'],\n",
    "    ['Fe3','cpd10516'],\n",
    "#     ['Sodium molybdate','cpd11145'], # This doesn't connect to anything\n",
    "    ['Ni2+','cpd00244'],\n",
    "    ['Selenate','cpd03396'],\n",
    "    ['Selenite','cpd03387'],\n",
    "    ['Zn2+','cpd00034']\n",
    "]\n",
    "\n",
    "# Enviromental Metabolites with Exchange reactions\n",
    "[\n",
    "#     ['CO2','cpd00011'], #EXs : \n",
    "#     ['Ca2+','cpd00063'], #EXs : \n",
    "#     ['Cd2+','cpd01012'], #EXs : Removed because toxic\n",
    "#     ['chromate','cpd11595'], #EXs : Removed because toxic\n",
    "#     ['Cl-','cpd00099'], #EXs : \n",
    "#     ['Co2+','cpd00149'], #EXs : In M9\n",
    "#     ['Cu2+','cpd00058'], #EXs : In M9\n",
    "#     ['Fe2+','cpd10515'], #EXs : \n",
    "#     ['H+','cpd00067'], #EXs : \n",
    "#     ['H2','cpd11640'], #EXs : \n",
    "#     ['H2O','cpd00001'], #EXs : \n",
    "#     ['Hg2+','cpd00531'], #EXs : Removed because toxic\n",
    "#     ['K+','cpd00205'], #EXs : \n",
    "#     ['Mg','cpd00254'], #EXs : \n",
    "#     ['Mn2+','cpd00030'], #EXs : \n",
    "#     ['Na+','cpd00971'], #EXs : \n",
    "#     ['Ni2+','cpd00244'], #EXs : In M9\n",
    "#     ['O2','cpd00007'], #EXs : \n",
    "#     ['Pb','cpd04097'], #EXs : Removed because toxic\n",
    "#     ['Zn2+','cpd00034'], #EXs : In M9\n",
    "#     ['fe3','cpd10516'] #EXs : In M9\n",
    "]\n",
    "\n",
    "# M9 Base : https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4932939/\n",
    "# [\n",
    "#     ['Ca2+','cpd00063'],\n",
    "#     ['Cl-','cpd00099'],\n",
    "#     ['CO2','cpd00011'],\n",
    "#     ['Co2+','cpd00149'],\n",
    "#     ['Cu2+','cpd00058'],\n",
    "#     ['Fe2+','cpd10515'],\n",
    "#     ['Fe3','cpd10516'],\n",
    "#     ['H+','cpd00067'],\n",
    "#     ['H2O','cpd00001'],\n",
    "#     ['K+','cpd00205'],\n",
    "#     ['Mg','cpd00254'],\n",
    "#     ['Mn2+','cpd00030'],\n",
    "#     ['Sodium molybdate','cpd11145'],\n",
    "#     ['Na+','cpd00971'],\n",
    "#     ['Ni2+','cpd00244'],\n",
    "#     ['Selenate','cpd03396'],\n",
    "#     ['Selenite','cpd03387'],\n",
    "#     ['Zn2+','cpd00034']\n",
    "# ]\n",
    "\n",
    "# M9 default carbon, nitrogen, phosphorous, and sulfur sources\n",
    "M9_sources = [\n",
    "    ['D-Glucose','cpd00027'],\n",
    "    ['NH3','cpd00013'], # this is actually NH4 : ammonium\n",
    "    ['Phosphate','cpd00009'],\n",
    "    ['Sulfate','cpd00048']\n",
    "]\n",
    "\n",
    "# Vitamins\n",
    "vit_k = [\n",
    "#     ['BIOT','cpd00104'], #EXs : Biotin\n",
    "#     ['Cobinamide','cpd03422'], #EXs : related to cobalamin (B12)\n",
    "#     ['Folate','cpd00393'], #EXs : \n",
    "    ['Menaquinone 7','cpd11606'], #EXs : Vitamine K2 : Add when there is no O2\n",
    "#     ['Niacin','cpd00218'], #EXs : \n",
    "#     ['PAN','cpd00644'], #EXs : Pantothenate\n",
    "#     ['Pyridoxal','cpd00215'], #EXs : \n",
    "#     ['Riboflavin','cpd00220'], #EXs : \n",
    "#     ['Thiamin','cpd00305'] #EXs : \n",
    "]\n",
    "\n",
    "# For aerobic simulations, O2 was added with a lower bound of âˆ’20 and to 0 for anaerobic simulations.\n",
    "\n",
    "# DNA/RNA related metabolites\n",
    "rna_bases = [\n",
    "#     ['35ccmp','cpd00696'], #EXs : \n",
    "#     ['AMP','cpd00018'], #EXs : \n",
    "    ['Adenosine','cpd00182'], #EXs : In BSM (as adenine)\n",
    "#     ['Adenosine 3-5-bisphosphate','cpd00045'], #EXs : \n",
    "    ['Cytosine','cpd00307'], #EXs : \n",
    "#     ['Deoxyadenosine','cpd00438'], #EXs : \n",
    "#     ['Deoxycytidine','cpd00654'], #EXs : \n",
    "#     ['Deoxyguanosine','cpd00277'], #EXs : In BSM\n",
    "#     ['Deoxyinosine','cpd03279'], #EXs : \n",
    "#     ['Deoxyuridine','cpd00412'], #EXs : \n",
    "#     ['GMP','cpd00126'], #EXs : \n",
    "#     ['GTP','cpd00038'], #EXs : \n",
    "    ['Guanosine','cpd00311'], #EXs : In BSM (as Guanine)\n",
    "#     ['Inosine','cpd00246'], #EXs : \n",
    "#     ['HYXN','cpd00226'], #EXs : Hypoxanthine\n",
    "#     ['Nicotinamide ribonucleotide','cpd00355'], #EXs : \n",
    "#     ['TTP','cpd00357'], #EXs : Deoxythymidine triphosphate\n",
    "    ['Thymidine','cpd00184'], #EXs : In BSM\n",
    "#     ['Thyminose','cpd01242'], #EXs : deoxyribose\n",
    "#     ['Uracil','cpd00092'], #EXs : \n",
    "    ['Uridine','cpd00249'], #EXs : In BSM (as uracil)\n",
    "#     ['XAN','cpd00309'], #EXs : Xanthine\n",
    "#     ['Xanthosine','cpd01217'], #EXs : \n",
    "#     ['dATP','cpd00115'], #EXs : \n",
    "#     ['dGTP','cpd00241'], #EXs : \n",
    "#     ['dTMP','cpd00298'] #EXs : \n",
    "]\n",
    "\n",
    "# Check to see if these metabolites are used in pathways? Should I add some of these to media? \n",
    "# Yes for ATP, and GTP. (TTP, CTP as well?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amino Acid related metabolites\n",
    "aas = [\n",
    "    ['D-Alanine','cpd00117'], #EXs : \n",
    "    ['D-Glutamate','cpd00186'], #EXs : \n",
    "    ['D-Methionine','cpd00637'], #EXs : \n",
    "    ['D-Serine','cpd00550'], #EXs : \n",
    "    ['Glycine','cpd00033'], #EXs : 1\n",
    "    ['L-Alanine','cpd00035'], #EXs : 2\n",
    "    ['L-Arginine','cpd00051'], #EXs : 3\n",
    "    ['L-Asparagine','cpd00132'], #EXs : 4\n",
    "    ['L-Aspartate','cpd00041'], #EXs : 5\n",
    "\n",
    "    ['L-Cysteine','cpd00084'], #EXs : 7\n",
    "    ['L-Glutamate','cpd00023'], #EXs : 8\n",
    "    ['L-Glutamine','cpd00053'], #EXs : 9\n",
    "    ['L-Histidine','cpd00119'], #EXs : 10\n",
    "    ['L-Isoleucine','cpd00322'], #EXs : 11\n",
    "    ['L-Leucine','cpd00107'], #EXs : 12\n",
    "    ['L-Lysine','cpd00039'], #EXs : 13\n",
    "    ['L-Methionine','cpd00060'], #EXs : 14\n",
    "    ['L-Phenylalanine','cpd00066'], #EXs : 15\n",
    "    ['L-Proline','cpd00129'], #EXs : 16\n",
    "    ['L-Serine','cpd00054'], #EXs : 17\n",
    "    ['L-Threonine','cpd00161'], #EXs : 18\n",
    "    ['L-Tryptophan','cpd00065'], #EXs : 19\n",
    "    ['L-Tyrosine','cpd00069'], #EXs : 20\n",
    "    ['L-Valine','cpd00156'] #EXs : 21\n",
    "]\n",
    "# Explore leave one out with core amino acids. \n",
    "\n",
    "# Dimers, and other amino acid related mets\n",
    "aa_related = [\n",
    "    ['2-Oxoglutarate','cpd00024'], #EXs : \n",
    "    ['Ala-Gln','cpd11587'], #EXs : \n",
    "    ['Ala-His','cpd11584'], #EXs : \n",
    "    ['Ala-Leu','cpd11583'], #EXs : \n",
    "    ['ala-L-asp-L','cpd11593'], #EXs : \n",
    "    ['ala-L-glu-L','cpd11586'], #EXs : \n",
    "    ['ala-L-Thr-L','cpd11582'], #EXs : \n",
    "    ['Aminoethanol','cpd00162'], #EXs : Ethanolamine\n",
    "    ['Carnitine','cpd00266'], #EXs : \n",
    "    ['Chorismate','cpd00216'], #EXs : \n",
    "    ['L-Cysteate','cpd00395'], #EXs : \n",
    "    ['Cys-Gly','cpd01017'], #EXs : \n",
    "    ['Gly-Cys','cpd15603'], #EXs : \n",
    "    ['Gly-Gln','cpd11580'], #EXs : \n",
    "    ['Gly-Leu','cpd15604'], #EXs : \n",
    "    ['Gly-Met','cpd11591'], #EXs : \n",
    "    ['Gly-Phe','cpd15605'], #EXs : \n",
    "    ['Gly-Tyr','cpd15606'], #EXs : \n",
    "    ['gly-asn-L','cpd11581'], #EXs : \n",
    "    ['gly-asp-L','cpd11589'], #EXs : \n",
    "    ['gly-glu-L','cpd11592'], #EXs : \n",
    "    ['gly-pro-L','cpd11588'], #EXs : \n",
    "    ['L-Methionine S-oxide','cpd01914'], #EXs :\n",
    "    ['L-alanylglycine','cpd11585'], #EXs : \n",
    "    ['L-methionine R-oxide','cpd11576'], #EXs : \n",
    "    ['met-L-ala-L','cpd11590'], #EXs :\n",
    "    ['S-Adenosyl-L-methionine','cpd00017'], #EXs : \n",
    "    ['S-Methyl-L-methionine','cpd02027'], #EXs : \n",
    "    ['S-Ribosylhomocysteine','cpd02227'], #EXs : \n",
    "    ['N-Acetyl-D-glucosamine','cpd00122'], #EXs : \n",
    "    ['N-Acetyl-D-mannosamine','cpd00492'], #EXs : \n",
    "    ['Ornithine','cpd00064'], #EXs : \n",
    "    ['Putrescine','cpd00118'], #EXs : \n",
    "    ['Taurine','cpd00210'], #EXs : \n",
    "    ['meso-2,6-Diaminopimelate','cpd00516'] #EXs : related to lysine\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Models...Starting Loop...\n",
      "Loop0 Optimizing...1000.0...Add to bag...Bag size: 1064...Add demand to model...\n",
      "1.0*DM_cpd00117_c - 1.0*DM_cpd00117_c_reverse_21bac...\n",
      "Penalties...Gapfilling...pFBA...1000.0...Constructing Dict...Ave likelihood of: cpd00117_c is 0.4436602411303152Resetting...Run time: 0.4274543643 [mins]\n",
      "Loop1 Optimizing...1000.0...Add to bag...Bag size: 1077...Add demand to model...\n",
      "1.0*DM_cpd00186_c - 1.0*DM_cpd00186_c_reverse_17bfd...\n",
      "Penalties...Gapfilling...pFBA...500.0...Constructing Dict...Ave likelihood of: cpd00186_c is 0.3938533157207162Resetting...Run time: 1.26339983145 [mins]\n",
      "Loop2 Optimizing...1000.0...Add to bag...Bag size: 1090...Add demand to model...\n",
      "1.0*DM_cpd00637_c - 1.0*DM_cpd00637_c_reverse_77fb4...\n",
      "Penalties...Gapfilling..."
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "failed to validate gapfilled model, try lowering the integer_threshold",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-2d998134ce4d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[0;31m#         bag.remove_reactions(reactions_to_remove)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m         \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Gapfilling...'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 92\u001b[0;31m         \u001b[0mgaps_to_fill\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgapfill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbag\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpenalties\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpenalties\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdemand_reactions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     93\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     94\u001b[0m     \u001b[0;31m# Fill the gaps\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/tjmoutinho/Lactobacillus/Code/gapfill2.pyc\u001b[0m in \u001b[0;36mgapfill\u001b[0;34m(model, universal, lower_bound, penalties, demand_reactions, exchange_reactions, iterations)\u001b[0m\n\u001b[1;32m    297\u001b[0m                           \u001b[0mdemand_reactions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdemand_reactions\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    298\u001b[0m                           exchange_reactions=exchange_reactions)\n\u001b[0;32m--> 299\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mgapfiller\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterations\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0miterations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/home/tjmoutinho/Lactobacillus/Code/gapfill2.pyc\u001b[0m in \u001b[0;36mfill\u001b[0;34m(self, iterations)\u001b[0m\n\u001b[1;32m    226\u001b[0m                         ind._get_primal() > self.integer_threshold]\n\u001b[1;32m    227\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msolution\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 228\u001b[0;31m                 raise RuntimeError('failed to validate gapfilled model, '\n\u001b[0m\u001b[1;32m    229\u001b[0m                                    'try lowering the integer_threshold')\n\u001b[1;32m    230\u001b[0m             \u001b[0mused_reactions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msolution\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: failed to validate gapfilled model, try lowering the integer_threshold"
     ]
    }
   ],
   "source": [
    "t = time.time()\n",
    "\n",
    "sys.stdout.write('Loading Models...')\n",
    "universal = cobra.io.load_json_model('universal.json')\n",
    "model = cobra.io.load_json_model('model.json')\n",
    "bag = cobra.io.load_json_model('bag.json')\n",
    "genome_id = '220668.9'\n",
    "likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "model.solver.configuration.verbosity = 3\n",
    "# bag.solver.configuration.verbosity = 3\n",
    "\n",
    "counter = 0\n",
    "\n",
    "total_dataset_dict = {}\n",
    "carb_idx = 0\n",
    "nit_idx = 0\n",
    "product_idx = 0\n",
    "carbon = 'D-Glucose'\n",
    "nitrogen = 'NH3'\n",
    "\n",
    "### Double For-loop to set media\n",
    "\n",
    "# Create specific Media List\n",
    "media_list = bsm + M9_sources + rna_bases # + nitrogen + carbon\n",
    "set_media(model, media_list, universal, verbose=False)\n",
    "set_media(universal, media_list, universal, verbose=False)\n",
    "\n",
    "# Run through each amino acid to check for production\n",
    "aa_like = {}\n",
    "sys.stdout.write('Starting Loop...')\n",
    "for aa_list in aas:\n",
    "    \n",
    "    sys.stdout.write('\\n'+ 'Loop' + str(counter) + ' ')\n",
    "    aa = aa_list[1]+'_c'\n",
    "    product = aa_list[0]\n",
    "    \n",
    "#     task_lb = 100\n",
    "    obj = 'DM_'+ aa\n",
    "    universal.objective = universal.reactions.get_by_id(obj)\n",
    "    \n",
    "    # Optimize universal with FBA to get possible reactions\n",
    "    sys.stdout.write('Optimizing...')\n",
    "    solution = universal.optimize()\n",
    "    sys.stdout.write(str(round(universal.slim_optimize())) + '...')\n",
    "    df = solution.fluxes.to_frame()\n",
    "    active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "    \n",
    "    # Add solution space to bag\n",
    "    sys.stdout.write('Add to bag...')\n",
    "    rxns_to_add_to_bag = []\n",
    "    for rxn in active.index:\n",
    "        if rxn not in [reaction.id for reaction in bag.reactions]:\n",
    "            rxns_to_add_to_bag.append(universal.reactions.get_by_id(rxn))\n",
    "    rxns_to_add_to_bag_copy = deepcopy(rxns_to_add_to_bag)\n",
    "    \n",
    "    bag.add_reactions(rxns_to_add_to_bag_copy)\n",
    "    \n",
    "    sys.stdout.write('Bag size: '+ str(len(bag.reactions)) +'...')\n",
    "    \n",
    "    # Add Demand Reaction for metabolite of interest and set to be objective\n",
    "    sys.stdout.write('Add demand to model...')\n",
    "    if aa not in [met.id for met in model.metabolites]:\n",
    "        metabolite = deepcopy(universal.metabolite.get_by_id(aa))\n",
    "        model.add_metabolites([metabolite])\n",
    "    else:\n",
    "        metabolite = model.metabolites.get_by_id(aa)\n",
    "    demand = model.add_boundary(metabolite, type='demand')\n",
    "#     model.add_reactions([demand])\n",
    "    model.objective = demand\n",
    "    sys.stdout.write('\\n' + str(model.objective.expression) + '...')\n",
    "    \n",
    "    # Gapfill with probanno\n",
    "    sys.stdout.write('\\nPenalties...')\n",
    "    # model.solver = 'gurobi'\n",
    "#     gaps_to_fill = probabilistic_gapfill(model, bag, likelihoods, dm_rxns=False, ex_rxns=False)\n",
    "#     gaps_to_fill = gapfill(model, bag, demand_reactions=False) # Update to probannopy gapfill function; use Gurobi\n",
    "    \n",
    "    with bag as bag:\n",
    "        default_penalties = {'Universal': 1, 'Exchange': 100, 'Demand': 1, 'Reverse': 75}\n",
    "        penalties = default_penalties\n",
    "        reactions_to_remove = []\n",
    "        for r in bag.reactions:\n",
    "            if model.reactions.has_id(r.id):\n",
    "                reactions_to_remove.append(r)\n",
    "                penalties[r.id] = 0  # In the model\n",
    "            elif r.id in likelihoods:\n",
    "                penalties[r.id] = max(0, 1 - likelihoods[r.id]) * (penalties[r.id] if r.id in penalties else 1)\n",
    "            else:\n",
    "                penalties[r.id] = 1\n",
    "#         bag.remove_reactions(reactions_to_remove)\n",
    "        sys.stdout.write('Gapfilling...')\n",
    "        dont_continue = 0\n",
    "        try:\n",
    "            gaps_to_fill = gapfill(model, bag, penalties=penalties, demand_reactions=False)\n",
    "        except:\n",
    "            try:\n",
    "                gaps_to_fill = gapfill(model, bag, penalties=penalties, demand_reactions=False)\n",
    "            except:\n",
    "                dont_continue = 1\n",
    "                pass\n",
    "    \n",
    "    if dont_continue == 0:\n",
    "        # Fill the gaps\n",
    "        rxns_to_add = []\n",
    "        for gap in gaps_to_fill:\n",
    "            model.add_reactions(gap)\n",
    "\n",
    "        # Optimize with filled pathway\n",
    "        sys.stdout.write('pFBA...')\n",
    "        solution = pfba(model, objective = demand)\n",
    "        sys.stdout.write(str(round(model.slim_optimize())) + '...')\n",
    "\n",
    "        sys.stdout.write('Constructing Dict...')\n",
    "        # Find reactions that carry flux\n",
    "        df = solution.fluxes.to_frame()\n",
    "        active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "\n",
    "        # Acquire likelihood scores for reactions that carry flux\n",
    "        flux_rxns = []\n",
    "        like_list = []\n",
    "        for rxn in list(active.index):\n",
    "            if rxn.startswith('rxn'):\n",
    "                try:\n",
    "                    flux_rxns.append([str(rxn),likelihoods[str(rxn)]])\n",
    "                    like_list.append(likelihoods[str(rxn)])\n",
    "                except:\n",
    "                    pass\n",
    "        avg_like = np.mean(like_list)\n",
    "        sys.stdout.write('Ave likelihood of: ' + aa + ' is ' + str(avg_like)) \n",
    "\n",
    "        counter += 1\n",
    "\n",
    "        report_dict = {}\n",
    "\n",
    "        report_dict['Model_ID'] = genome_id\n",
    "        report_dict['Carbon'] = carbon\n",
    "        report_dict['Nitrogen'] = nitrogen\n",
    "        report_dict['objective'] = product\n",
    "        report_dict['avg_path_like'] = avg_like\n",
    "        report_dict['gaps_filled'] = gaps_to_fill[0]\n",
    "        report_dict['reactions_w_flux'] = flux_rxns\n",
    "        report_dict['active_rxns'] = active\n",
    "\n",
    "        report_dict_ID = genome_id + ':' + str(carb_idx) + '.' + str(nit_idx) + '.' + str(product_idx)\n",
    "        total_dataset_dict[report_dict_ID] = report_dict\n",
    "        product_idx += 1 #Keep track to which product is being maximized\n",
    "\n",
    "        # Remove reactions to reset\n",
    "        sys.stdout.write('Resetting...')\n",
    "        model.remove_reactions([demand])\n",
    "        model.remove_reactions(gaps_to_fill[0])\n",
    "        bag.remove_reactions(rxns_to_add_to_bag_copy)\n",
    "\n",
    "        elapsed = time.time() - t\n",
    "        sys.stdout.write('Run time: ' + str(elapsed/60) + \" [mins]\")\n",
    "\n",
    "    elif dont_continue = 1:\n",
    "        counter += 1\n",
    "\n",
    "        report_dict = {}\n",
    "\n",
    "        report_dict['Model_ID'] = genome_id\n",
    "        report_dict['Carbon'] = carbon\n",
    "        report_dict['Nitrogen'] = nitrogen\n",
    "        report_dict['objective'] = product\n",
    "        report_dict['avg_path_like'] = \"Failed to gapfill\"\n",
    "        report_dict['gaps_filled'] = \"Failed to gapfill\"\n",
    "        report_dict['reactions_w_flux'] = \"Failed to gapfill\"\n",
    "        report_dict['active_rxns'] = \"Failed to gapfill\"\n",
    "\n",
    "        report_dict_ID = genome_id + ':' + str(carb_idx) + '.' + str(nit_idx) + '.' + str(product_idx)\n",
    "        total_dataset_dict[report_dict_ID] = report_dict\n",
    "        product_idx += 1 #Keep track to which product is being maximized\n",
    "\n",
    "        # Remove reactions to reset\n",
    "        sys.stdout.write('Resetting...')\n",
    "        model.remove_reactions([demand])\n",
    "        bag.remove_reactions(rxns_to_add_to_bag_copy)\n",
    "\n",
    "        elapsed = time.time() - t\n",
    "        sys.stdout.write('Run time: ' + str(elapsed/60) + \" [mins]\")\n",
    "        \n",
    "        \n",
    "file_name = \"../metabolic_output/%s.data\" % (genome_id)\n",
    "pickle.dump(total_dataset_dict, open(file_name, \"wb\"))\n",
    "\n",
    "elapsed = time.time() - t\n",
    "print(\"\\nTime to complete: \" + str(elapsed/60) + \" [mins]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_dataset_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pFBA with new media components (RNA bases + thymidine...) \n",
    "# Remove reaction likelihoods of zero from model\n",
    "# Add demands for all metabolites in model to avoid any reactions being blocked\n",
    "# Use thresholded likelihoods to penalize reactions with no likelihood using pFBA gapfill\n",
    "# Use pFBA answer + 0 likelihood reactions from reconstructed model + higher likelihood reactions to probanno gapfill\n",
    "\n",
    "t = time.time()\n",
    "counter = 0\n",
    "\n",
    "sys.stdout.write('Loading in models...')\n",
    "\n",
    "universal = cobra.io.load_json_model(\"../Data/GramPosUni.json\")\n",
    "genome_id = '220668.9'\n",
    "model = cobra.io.read_sbml_model('../gap_models/'+ genome_id +'.xml')\n",
    "likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "\n",
    "sys.stdout.write('Adding Water...')\n",
    "\n",
    "# Ensure free diffusion of water\n",
    "model.reactions.get_by_id('rxn05319_c').name = \"Water transport\"\n",
    "model.reactions.get_by_id('rxn05319_c').bounds = (-1000., 1000.)\n",
    "\n",
    "sys.stdout.write('Set-up Universal...')\n",
    "\n",
    "### Set up Universal\n",
    "# Add all model reactions to Universal model\n",
    "model_reactions = copy.deepcopy(model.reactions)\n",
    "model_reactions_to_remove = []\n",
    "for rxn in model.reactions:\n",
    "    if rxn.id in set([reaction.id for reaction in universal.reactions]):\n",
    "        model_reactions_to_remove.append(rxn)\n",
    "universal.remove_reactions([reaction.id for reaction in model_reactions_to_remove])\n",
    "universal.add_reactions(model_reactions)\n",
    "\n",
    "# Add GPRs from likelihood dict to universal reactions that are still missing them\n",
    "rxn_ids = [reaction.id for reaction in universal.reactions]\n",
    "rxns_to_fix = []\n",
    "for rxn in rxn_ids:\n",
    "    if rxn.startswith('rxn'):\n",
    "        try: # This catches all of the reactions that have no likelihood value.\n",
    "            if likelihoods[rxn] > 0.0:\n",
    "                if universal.reactions.get_by_id(rxn).gene_reaction_rule == '':\n",
    "                    if likelihoods.data[rxn]['gpr'] != '':\n",
    "                        rxns_to_fix.append(rxn)\n",
    "                        universal.reactions.get_by_id(rxn).gene_reaction_rule = likelihoods.data[rxn]['gpr']\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "# Add demand for all metabolites in Universal model to stop blocked reactions\n",
    "all_mets = []\n",
    "for met in universal.metabolites:\n",
    "    if (met.id.endswith('_c')):\n",
    "        universal.add_boundary(met, type='demand')\n",
    "\n",
    "### Set Up Model: remove low likelihood reactions\n",
    "sys.stdout.write('Set-up Model...')\n",
    "low_like_model = []\n",
    "for rxn in model.reactions:\n",
    "    if rxn.id.startswith('rxn'):\n",
    "        try:\n",
    "            if likelihoods[rxn.id] <= 0.1:\n",
    "                low_like_model.append(rxn.id)\n",
    "        except:\n",
    "            pass\n",
    "model_rxns_to_remove = [model.reactions.get_by_id(rxn) for rxn in low_like_model]\n",
    "model.remove_reactions(model_rxns_to_remove)\n",
    "\n",
    "### Set Up Bag: Leave only low-likelihood-model reactions, and high-likelihood-non-model reactions for now\n",
    "sys.stdout.write('Set-up Bag...')\n",
    "\n",
    "# Make deepcopy of universal for bag to process later\n",
    "bag = copy.deepcopy(universal)\n",
    "\n",
    "# Find reaction IDs for the high-likelihood reactions only in the universal model\n",
    "high_like_non_model = []\n",
    "for rxn in universal.reactions:\n",
    "    if rxn.id.startswith('rxn') and rxn.id not in [reaction.id for reaction in model.reactions]:\n",
    "        try:\n",
    "            if likelihoods[rxn.id] >= 0.1:\n",
    "                high_like_non_model.append(rxn.id)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "uni_rxn_ids = [reaction.id for reaction in universal.reactions]\n",
    "rxns_to_remove = set(uni_rxn_ids).difference(set(high_like_non_model).union(set(low_like_model)))\n",
    "\n",
    "sys.stdout.write('Trimming Bag...')\n",
    "bag.remove_reactions([bag.reactions.get_by_id(rxn) for rxn in rxns_to_remove])\n",
    "\n",
    "# Save Models\n",
    "sys.stdout.write('Saving...')\n",
    "cobra.io.save_json_model(universal, \"universal.json\")\n",
    "cobra.io.save_json_model(model, \"model.json\")\n",
    "cobra.io.save_json_model(bag, \"bag.json\")\n",
    "\n",
    "elapsed = time.time() - t\n",
    "print(\"\\nTime to complete: \" + str(elapsed/60) + \" [mins]\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
