{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "import cobra\n",
    "import cobra.test\n",
    "# import mackinac\n",
    "import numpy as np\n",
    "import csv\n",
    "import glob\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import time\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "from cobra.flux_analysis import gapfill\n",
    "from cobra.flux_analysis import pfba\n",
    "\n",
    "# Set default logger to python logger to avoid warnings given when adding reactions and/or metaboites \n",
    "# because \"cobra.core.model\" doesn't innately have a logger.\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger('logger')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_media(model, media, universal, verbose=False):\n",
    "\n",
    "    # Find and close all exchange reactions in the model\n",
    "    model_rxns = [rxn.id for rxn in model.reactions]\n",
    "    for rxn in model_rxns:\n",
    "        if rxn.startswith('EX_') and rxn.endswith('_e'):\n",
    "            model.reactions.get_by_id(rxn).lower_bound = 0.0\n",
    "\n",
    "    # Check for existence of exchange reactions for the media metabolites in the model\n",
    "    for metabolite in media:\n",
    "        met = metabolite[1]+'_e'\n",
    "        if 'EX_'+met in model_rxns:\n",
    "            model.reactions.get_by_id('EX_'+met).lower_bound = -1000.\n",
    "        else:\n",
    "            # Create exchange reaction and add to model\n",
    "            if verbose:\n",
    "                print(\"added exchange rxn for \" + met)\n",
    "            new_exchange = cobra.Reaction('EX_'+met)\n",
    "            new_exchange.name = met + ' exchange'\n",
    "            met_obj = universal.metabolites.get_by_id(met)\n",
    "            new_exchange.add_metabolites({met_obj:-1})\n",
    "            new_exchange.lower_bound = -1000.\n",
    "            new_exchange.upper_bound = 1000.\n",
    "            model.add_reaction(new_exchange)\n",
    "            model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in list of genome IDs\n",
    "with open('../Data/20_species_1023_genomes.csv') as csvfile:\n",
    "    genome_ids_list = []\n",
    "    for line in csvfile:\n",
    "        genome_ids_list.append(line.strip())\n",
    "len(genome_ids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize functions\n",
    "# Initialize global variables\n",
    "# Set media cpds\n",
    "# Open model\n",
    "# Add water transporter and fix name\n",
    "# loop: Change media\n",
    "    # turn off all exchanges\n",
    "    # Turn on correct exchanges, add if missing\n",
    "    \n",
    "    # Loop: Set demand reaction\n",
    "        # Check for production\n",
    "        # Gapfill with probanno\n",
    "        # Determine the reactions that were added to gapfill and identify the likelihoods\n",
    "        # Optimize and find solution\n",
    "        # Average likelihoods to create metric for media condition and specific demand reaction\n",
    "        # save information in data structure\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting Loop\n",
      "0...gapfilling...optimizing...Average Likelihood of: cpd00117 is 0.32239418829432986\n",
      "1...gapfilling...optimizing...Average Likelihood of: cpd00186 is 0.40811404675637747\n",
      "2...gapfilling...optimizing...Average Likelihood of: cpd00637 is 0.3539055920939708\n",
      "3...gapfilling...optimizing...Average Likelihood of: cpd00550 is 0.3876222618910973\n",
      "4...gapfilling...optimizing...Average Likelihood of: cpd00033 is 0.25864420133992655\n",
      "5...gapfilling...optimizing...Average Likelihood of: cpd00035 is 0.2617434799782508\n",
      "6...gapfilling...optimizing...Average Likelihood of: cpd00051 is 0.30615169412786564\n",
      "7...gapfilling...optimizing...Average Likelihood of: cpd00132 is 0.37908217357377966\n",
      "8...gapfilling...optimizing...Average Likelihood of: cpd00041 is 0.3919271330103651\n",
      "9...gapfilling..."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No handlers could be found for logger \"cobra.core.model\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimizing...Average Likelihood of: cpd00084 is 0.40759115411391483\n",
      "10...gapfilling...optimizing...Average Likelihood of: cpd00023 is 0.3770658617850227\n",
      "11...gapfilling...optimizing...Average Likelihood of: cpd00053 is 0.24679149221825047\n",
      "12...gapfilling...optimizing...Average Likelihood of: cpd00119 is 0.34746531593017904\n",
      "13...gapfilling...optimizing...Average Likelihood of: cpd00322 is 0.3048953761712908\n",
      "14...gapfilling...optimizing...Average Likelihood of: cpd00107 is 0.35867506074632505\n",
      "15...gapfilling...optimizing...Average Likelihood of: cpd00039 is 0.38470369962458417\n",
      "16...gapfilling...optimizing...Average Likelihood of: cpd00060 is 0.3625340412746876\n",
      "17...gapfilling...optimizing...Average Likelihood of: cpd00066 is 0.37888939227146257\n",
      "18...gapfilling...optimizing...Average Likelihood of: cpd00129 is 0.3553058212741118\n",
      "19...gapfilling...optimizing...Average Likelihood of: cpd00054 is 0.2855063409092786\n",
      "20...gapfilling...optimizing...Average Likelihood of: cpd00161 is 0.396418896014934\n",
      "21...gapfilling...optimizing...Average Likelihood of: cpd00065 is 0.38664620586239773\n",
      "22...gapfilling...optimizing...Average Likelihood of: cpd00069 is 0.3976185204163214\n",
      "23...gapfilling...optimizing...Average Likelihood of: cpd00156 is 0.317618970729169Time to complete:505.108682565mins\n"
     ]
    }
   ],
   "source": [
    "t = time.time()\n",
    "counter = 0\n",
    "\n",
    "universal = cobra.io.load_json_model(\"../Data/GramPosUni.json\")\n",
    "genome_id = '220668.9'\n",
    "model = cobra.io.read_sbml_model('../gap_models/'+ genome_id +'.xml')\n",
    "likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "\n",
    "# Ensure free water exhange\n",
    "model.reactions.get_by_id('rxn05319_c').name = \"Water transport\"\n",
    "model.reactions.get_by_id('rxn05319_c').bounds = (-1000., 1000.)\n",
    "\n",
    "# Create specific Media List\n",
    "media_list = bsm + M9_sources\n",
    "set_media(model, media_list, universal, verbose=False)\n",
    "\n",
    "# Run through each amino acid to check for production\n",
    "aa_like = {}\n",
    "sys.stdout.write('Starting Loop')\n",
    "for aa_list in aas:\n",
    "    sys.stdout.write('\\n'+ str(counter))\n",
    "    aa = aa_list[1]+'_c'\n",
    "    # Add Demand Reaction for metabolite\n",
    "    metabolite = model.metabolites.get_by_id(aa)\n",
    "    demand = model.add_boundary(metabolite, type='demand')\n",
    "    model.objective = demand\n",
    "    # Gapfill\n",
    "    sys.stdout.write('...gapfilling...')\n",
    "    gaps_to_fill = gapfill(model, universal, demand_reactions=False) # Update to probannopy gapfill function; use Gurobi\n",
    "    # Fill the gaps\n",
    "    rxns_to_add = []\n",
    "    for gap in gaps_to_fill:\n",
    "        model.add_reactions(gap) ### I NEED TO REMOVE THESE REACTIONS\n",
    "    # Optimize with full pathway\n",
    "    sys.stdout.write('optimizing...')\n",
    "    solution = model.optimize()\n",
    "    # Find reactions that carry flux and their likelihood score\n",
    "    df = solution.fluxes.to_frame()\n",
    "    active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "    like_list = []\n",
    "    for rxn in list(active.index):\n",
    "        if rxn.startswith('rxn'):\n",
    "            try:\n",
    "                like_list.append(likelihoods[rxn])\n",
    "            except:\n",
    "                pass\n",
    "    avg_like = np.mean(like_list)\n",
    "    sys.stdout.write('Average Likelihood of: ' + aa_list[1] + ' is ' + str(avg_like))\n",
    "    aa_like[aa_list[1]] = avg_like\n",
    "    model.remove_reactions([demand])\n",
    "    # ADD REMOVAL OF GAPS\n",
    "    counter += 1\n",
    "\n",
    "elapsed = time.time() - t\n",
    "print(\"Time to complete:\" + str(elapsed/60) + \"mins\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cpd00023': 0.3770658617850227,\n",
       " 'cpd00033': 0.25864420133992655,\n",
       " 'cpd00035': 0.2617434799782508,\n",
       " 'cpd00039': 0.38470369962458417,\n",
       " 'cpd00041': 0.3919271330103651,\n",
       " 'cpd00051': 0.30615169412786564,\n",
       " 'cpd00053': 0.24679149221825047,\n",
       " 'cpd00054': 0.2855063409092786,\n",
       " 'cpd00060': 0.3625340412746876,\n",
       " 'cpd00065': 0.38664620586239773,\n",
       " 'cpd00066': 0.37888939227146257,\n",
       " 'cpd00069': 0.3976185204163214,\n",
       " 'cpd00084': 0.40759115411391483,\n",
       " 'cpd00107': 0.35867506074632505,\n",
       " 'cpd00117': 0.32239418829432986,\n",
       " 'cpd00119': 0.34746531593017904,\n",
       " 'cpd00129': 0.3553058212741118,\n",
       " 'cpd00132': 0.37908217357377966,\n",
       " 'cpd00156': 0.317618970729169,\n",
       " 'cpd00161': 0.396418896014934,\n",
       " 'cpd00186': 0.40811404675637747,\n",
       " 'cpd00322': 0.3048953761712908,\n",
       " 'cpd00550': 0.3876222618910973,\n",
       " 'cpd00637': 0.3539055920939708}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There are no demand or exchange reactions in the universal reaction bag\n",
    "for rxn in universal.reactions:\n",
    "    if rxn.id.startswith('DM'):\n",
    "        print('DM in reaction bag')\n",
    "    elif rxn.id.startswith('EX'):\n",
    "        print('EX in reaction bag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[<Reaction rxn00898_c at 0x7eff87546350>,\n",
       "  <Reaction rxn02186_c at 0x7eff87546410>]]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaps_to_fill"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "print(likelihoods['rxn00898_c'])\n",
    "print(likelihoods['rxn02186_c'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.2733090489303243\n",
      "975\n",
      "18\n"
     ]
    }
   ],
   "source": [
    "all_likes = []\n",
    "passed_rxns = []\n",
    "counter = 0\n",
    "pass_counter = 0\n",
    "for rxn in model.reactions:\n",
    "    if rxn.id.startswith('rxn'):\n",
    "        try:\n",
    "            all_likes.append(likelihoods[rxn.id])\n",
    "            counter += 1\n",
    "        except:\n",
    "            passed_rxns.append(rxn.id)\n",
    "            pass_counter += 1\n",
    "            pass\n",
    "global_avg = np.mean(all_likes)\n",
    "\n",
    "print(global_avg)\n",
    "print(counter)\n",
    "print(pass_counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rxn02374_c',\n",
       " 'rxn05319_c',\n",
       " 'rxn04457_c',\n",
       " 'rxn02916_c',\n",
       " 'rxn03012_c',\n",
       " 'rxn10571_c',\n",
       " 'rxn04132_c',\n",
       " 'rxn05195_c',\n",
       " 'rxn05468_c',\n",
       " 'rxn04133_c',\n",
       " 'rxn05467_c',\n",
       " u'rxn12215_c',\n",
       " u'rxn05522_c',\n",
       " u'rxn08688_c',\n",
       " u'rxn05238_c',\n",
       " u'rxn02976_c',\n",
       " u'rxn08764_c',\n",
       " u'rxn13022_c']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "passed_rxns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R03314\n",
      "Water transport\n",
      "R06605\n",
      "R04175\n",
      "R04336\n",
      "Mg2+-importing ATPase\n",
      "R06063\n",
      "ATP phosphohydrolase (ferric-ion-transporting)\n",
      "TRANS-RXNAVI-26568.ce\n",
      "R06064\n",
      "CO2 transporter via diffusion\n",
      "5-methyltetrahydropteroyltri-l-glutamate synthesis\n",
      "citrate transport in via Co complex\n",
      "Hydrogen sulfide oxidation\n",
      "TRANS-RXNBWI-115353.ce\n",
      "R04269\n",
      "ketol-acid reductoisomerase (2-Acetolactate)\n",
      "3-isopropylmalate dehydrogenase\n"
     ]
    }
   ],
   "source": [
    "print(model.reactions.get_by_id('rxn02374_c').name)\n",
    "print(model.reactions.get_by_id('rxn05319_c').name)\n",
    "print(model.reactions.get_by_id('rxn04457_c').name)\n",
    "print(model.reactions.get_by_id('rxn02916_c').name)\n",
    "print(model.reactions.get_by_id('rxn03012_c').name)\n",
    "print(model.reactions.get_by_id('rxn10571_c').name)\n",
    "print(model.reactions.get_by_id('rxn04132_c').name)\n",
    "print(model.reactions.get_by_id('rxn05195_c').name)\n",
    "print(model.reactions.get_by_id('rxn05468_c').name)\n",
    "print(model.reactions.get_by_id('rxn04133_c').name)\n",
    "print(model.reactions.get_by_id('rxn05467_c').name)\n",
    "print(model.reactions.get_by_id('rxn12215_c').name)\n",
    "print(model.reactions.get_by_id('rxn05522_c').name)\n",
    "print(model.reactions.get_by_id('rxn08688_c').name)\n",
    "print(model.reactions.get_by_id('rxn05238_c').name)\n",
    "print(model.reactions.get_by_id('rxn02976_c').name)\n",
    "print(model.reactions.get_by_id('rxn08764_c').name)\n",
    "print(model.reactions.get_by_id('rxn13022_c').name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R03314\n",
      "R06605\n",
      "R04175\n",
      "R04336\n",
      "R06063\n",
      "TRANS-RXNAVI-26568.ce\n",
      "R06064\n",
      "TRANS-RXNBWI-115353.ce\n",
      "R04269\n"
     ]
    }
   ],
   "source": [
    "print(model.reactions.get_by_id('rxn02374_c').name)\n",
    "# print(model.reactions.get_by_id('rxn05319_c').name)\n",
    "print(model.reactions.get_by_id('rxn04457_c').name)\n",
    "print(model.reactions.get_by_id('rxn02916_c').name)\n",
    "print(model.reactions.get_by_id('rxn03012_c').name)\n",
    "# print(model.reactions.get_by_id('rxn10571_c').name)\n",
    "print(model.reactions.get_by_id('rxn04132_c').name)\n",
    "# print(model.reactions.get_by_id('rxn05195_c').name)\n",
    "print(model.reactions.get_by_id('rxn05468_c').name)\n",
    "print(model.reactions.get_by_id('rxn04133_c').name)\n",
    "# print(model.reactions.get_by_id('rxn05467_c').name)\n",
    "# print(model.reactions.get_by_id('rxn12215_c').name)\n",
    "# print(model.reactions.get_by_id('rxn05522_c').name)\n",
    "# print(model.reactions.get_by_id('rxn08688_c').name)\n",
    "print(model.reactions.get_by_id('rxn05238_c').name)\n",
    "print(model.reactions.get_by_id('rxn02976_c').name)\n",
    "# print(model.reactions.get_by_id('rxn08764_c').name)\n",
    "# print(model.reactions.get_by_id('rxn13022_c').name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn02374_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R03314</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa8ae0810</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00858_c --> cpd00001_c + cpd00067_c + cpd02431_c</p>\n",
       "                    <p style='text-align:right'>L-Glutamate5-semialdehyde --> H2O + H+ + 1-Pyrroline-5-carboxylate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn02374_c at 0x7effa8ae0810>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn02374_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn04457_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R06605</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa4ffb350</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd09027_c --> cpd00011_c + cpd01567_c</p>\n",
       "                    <p style='text-align:right'>5-Hydroxy-2-oxo-4-ureido-2,5-dihydro-1H-imidazole-5-carboxylate --> CO2 + (R)-Allantoin</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn04457_c at 0x7effa4ffb350>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn04457_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn02916_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R04175</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa913bd50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd02074_c --> cpd00001_c + cpd00067_c + cpd00922_c</p>\n",
       "                    <p style='text-align:right'>2-Oxo-6-aminocaproate --> H2O + H+ + delta1-Piperideine-2-carboxylate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn02916_c at 0x7effa913bd50>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn02916_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn03012_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R04336</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa94f68d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd02414_c --> cpd00001_c + cpd00067_c + cpd02465_c</p>\n",
       "                    <p style='text-align:right'>L-2-Amino-6-oxopimelate --> H2O + H+ + tetrahydrodipicolinate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn03012_c at 0x7effa94f68d0>"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn03012_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn04132_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R06063</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa8c2a8d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00001_c + cpd08625_c --> cpd00011_c + cpd00067_c + cpd01567_c</p>\n",
       "                    <p style='text-align:right'>H2O + 5-Hydroxyisourate --> CO2 + H+ + (R)-Allantoin</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn04132_c at 0x7effa8c2a8d0>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn04132_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn05468_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>TRANS-RXNAVI-26568.ce</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effab166150</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00007_e <=> cpd00007_c</p>\n",
       "                    <p style='text-align:right'>O2 <=> O2</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn05468_c at 0x7effab166150>"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn05468_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn04133_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R06064</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effaabbc5d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00001_c + cpd08625_c --> cpd00011_c + cpd00067_c + cpd01092_c</p>\n",
       "                    <p style='text-align:right'>H2O + 5-Hydroxyisourate --> CO2 + H+ + Allantoin</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn04133_c at 0x7effaabbc5d0>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn04133_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn05238_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>TRANS-RXNBWI-115353.ce</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa33fb090</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00048_e <=> cpd00048_c</p>\n",
       "                    <p style='text-align:right'>Sulfate <=> Sulfate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn05238_c at 0x7effa33fb090>"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn05238_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>rxn02976_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R04269</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07eff9658de90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00024_c + cpd02273_c <=> cpd00023_c + cpd03591_c</p>\n",
       "                    <p style='text-align:right'>2-Oxoglutarate + L-threo-3-Methylaspartate <=> L-Glutamate + Methyloxaloacetate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction rxn02976_c at 0x7eff9658de90>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('rxn02976_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# turn off O2 exchange\n",
    "model.reactions.get_by_id('EX_cpd00007_e').lower_bound = 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basal Synthetic Media\n",
    "bsm = [\n",
    "    ['H+','cpd00067'],\n",
    "    ['H2O','cpd00001'],\n",
    "    ['CO2','cpd00011'],\n",
    "    ['O2','cpd00007'],\n",
    "    ['N2','cpd00528'], \n",
    "#     ['H2','cpd11640'], # Only with no O2\n",
    "    \n",
    "    ['K+','cpd00205'],\n",
    "    ['Na+','cpd00971'],\n",
    "    ['Mg','cpd00254'],\n",
    "    ['Mn2+','cpd00030'],\n",
    "    ['Fe2+','cpd10515'], # Iron ion in heme\n",
    "    ['Ca2+','cpd00063'], # Calcium pantothenate;cpd19112\n",
    "    \n",
    "    ['Vitamin B12r','cpd00423'], # C62H91CoN13O14P : cobalamin;cpd03424;cpd00730 : not present in any exchange reactions\n",
    "    ['Cobinamide','cpd03422'], #EXs : related to cobalamin (B12) Added to ensure cells have access to B12\n",
    "    ['BIOT','cpd00104'], # C10H15N2O3S : biotin B7\n",
    "    ['PAN','cpd00644'], # C9H16NO5 : Pantothenate B5\n",
    "    ['Folate','cpd00393'], # C19H17N7O6 : B9\n",
    "    ['Niacin','cpd00218'], # C6H4NO2 : B3\n",
    "    ['Pyridoxal','cpd00215'], # C8H9NO3 : B6\n",
    "    ['Riboflavin','cpd00220'], # C17H19N4O6 : B2\n",
    "    ['thiamin','cpd00305'], # C12H17N4OS : B1\n",
    "    \n",
    "#     ['Phosphate','cpd00009'], # HO4P : In M9 Defaults\n",
    "    \n",
    "    ['Thioglycolate','cpd01415'], # C2H3O2S : not present in any exchange reactions\n",
    "#     ['Sulfate','cpd00048'], # O4S : In M9 Defaults\n",
    "    \n",
    "    ['Acetate','cpd00029'], # C2H3O2 : not present in any exchange reactions\n",
    "    ['Citrate','cpd00137'], # C6H5O7 : Consider removing. \n",
    "#     ['Polysorbate 60','cpd24450'], # C35H68O10 : Almost tween 80 : not present in any reactions\n",
    "#     ['Ethyl acetate','cpd00633'], # C4H8O2 : not present in any exchange reactions, only present in one reaction at all\n",
    "    \n",
    "    ['ABEE','cpd00443'] # C7H6NO2 : aminobenzoate : not present in any exchange reactions\n",
    "]\n",
    "\n",
    "# Potentially add to BSM (from M9 media)\n",
    "M9_ions = [\n",
    "    ['Cl-','cpd00099'],\n",
    "    ['Co2+','cpd00149'],\n",
    "    ['Cu2+','cpd00058'],\n",
    "    ['Fe3','cpd10516'],\n",
    "#     ['Sodium molybdate','cpd11145'], # This doesn't connect to anything\n",
    "    ['Ni2+','cpd00244'],\n",
    "    ['Selenate','cpd03396'],\n",
    "    ['Selenite','cpd03387'],\n",
    "    ['Zn2+','cpd00034']\n",
    "]\n",
    "\n",
    "# Enviromental Metabolites with Exchange reactions\n",
    "[\n",
    "#     ['CO2','cpd00011'], #EXs : \n",
    "#     ['Ca2+','cpd00063'], #EXs : \n",
    "#     ['Cd2+','cpd01012'], #EXs : Removed because toxic\n",
    "#     ['chromate','cpd11595'], #EXs : Removed because toxic\n",
    "#     ['Cl-','cpd00099'], #EXs : \n",
    "#     ['Co2+','cpd00149'], #EXs : In M9\n",
    "#     ['Cu2+','cpd00058'], #EXs : In M9\n",
    "#     ['Fe2+','cpd10515'], #EXs : \n",
    "#     ['H+','cpd00067'], #EXs : \n",
    "#     ['H2','cpd11640'], #EXs : \n",
    "#     ['H2O','cpd00001'], #EXs : \n",
    "#     ['Hg2+','cpd00531'], #EXs : Removed because toxic\n",
    "#     ['K+','cpd00205'], #EXs : \n",
    "#     ['Mg','cpd00254'], #EXs : \n",
    "#     ['Mn2+','cpd00030'], #EXs : \n",
    "#     ['Na+','cpd00971'], #EXs : \n",
    "#     ['Ni2+','cpd00244'], #EXs : In M9\n",
    "#     ['O2','cpd00007'], #EXs : \n",
    "#     ['Pb','cpd04097'], #EXs : Removed because toxic\n",
    "#     ['Zn2+','cpd00034'], #EXs : In M9\n",
    "#     ['fe3','cpd10516'] #EXs : In M9\n",
    "]\n",
    "\n",
    "# M9 Base : https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4932939/\n",
    "# [\n",
    "#     ['Ca2+','cpd00063'],\n",
    "#     ['Cl-','cpd00099'],\n",
    "#     ['CO2','cpd00011'],\n",
    "#     ['Co2+','cpd00149'],\n",
    "#     ['Cu2+','cpd00058'],\n",
    "#     ['Fe2+','cpd10515'],\n",
    "#     ['Fe3','cpd10516'],\n",
    "#     ['H+','cpd00067'],\n",
    "#     ['H2O','cpd00001'],\n",
    "#     ['K+','cpd00205'],\n",
    "#     ['Mg','cpd00254'],\n",
    "#     ['Mn2+','cpd00030'],\n",
    "#     ['Sodium molybdate','cpd11145'],\n",
    "#     ['Na+','cpd00971'],\n",
    "#     ['Ni2+','cpd00244'],\n",
    "#     ['Selenate','cpd03396'],\n",
    "#     ['Selenite','cpd03387'],\n",
    "#     ['Zn2+','cpd00034']\n",
    "# ]\n",
    "\n",
    "# M9 default carbon, nitrogen, phosphorous, and sulfur sources\n",
    "M9_sources = [\n",
    "    ['D-Glucose','cpd00027'],\n",
    "    ['NH3','cpd00013'], # this is actually NH4 : ammonium\n",
    "    ['Phosphate','cpd00009'],\n",
    "    ['Sulfate','cpd00048']\n",
    "]\n",
    "\n",
    "# Vitamins\n",
    "vit_k = [\n",
    "#     ['BIOT','cpd00104'], #EXs : Biotin\n",
    "#     ['Cobinamide','cpd03422'], #EXs : related to cobalamin (B12)\n",
    "#     ['Folate','cpd00393'], #EXs : \n",
    "    ['Menaquinone 7','cpd11606'], #EXs : Vitamine K2 : Add when there is no O2\n",
    "#     ['Niacin','cpd00218'], #EXs : \n",
    "#     ['PAN','cpd00644'], #EXs : Pantothenate\n",
    "#     ['Pyridoxal','cpd00215'], #EXs : \n",
    "#     ['Riboflavin','cpd00220'], #EXs : \n",
    "#     ['Thiamin','cpd00305'] #EXs : \n",
    "]\n",
    "\n",
    "# For aerobic simulations, O2 was added with a lower bound of −20 and to 0 for anaerobic simulations.\n",
    "\n",
    "# DNA/RNA related metabolites\n",
    "rna_bases = [\n",
    "#     ['35ccmp','cpd00696'], #EXs : \n",
    "#     ['AMP','cpd00018'], #EXs : \n",
    "    ['Adenosine','cpd00182'], #EXs : In BSM (as adenine)\n",
    "#     ['Adenosine 3-5-bisphosphate','cpd00045'], #EXs : \n",
    "    ['Cytosine','cpd00307'], #EXs : \n",
    "#     ['Deoxyadenosine','cpd00438'], #EXs : \n",
    "#     ['Deoxycytidine','cpd00654'], #EXs : \n",
    "#     ['Deoxyguanosine','cpd00277'], #EXs : In BSM\n",
    "#     ['Deoxyinosine','cpd03279'], #EXs : \n",
    "#     ['Deoxyuridine','cpd00412'], #EXs : \n",
    "#     ['GMP','cpd00126'], #EXs : \n",
    "#     ['GTP','cpd00038'], #EXs : \n",
    "    ['Guanosine','cpd00311'], #EXs : In BSM (as Guanine)\n",
    "#     ['Inosine','cpd00246'], #EXs : \n",
    "#     ['HYXN','cpd00226'], #EXs : Hypoxanthine\n",
    "#     ['Nicotinamide ribonucleotide','cpd00355'], #EXs : \n",
    "#     ['TTP','cpd00357'], #EXs : Deoxythymidine triphosphate\n",
    "    ['Thymidine','cpd00184'], #EXs : In BSM\n",
    "#     ['Thyminose','cpd01242'], #EXs : deoxyribose\n",
    "#     ['Uracil','cpd00092'], #EXs : \n",
    "    ['Uridine','cpd00249'], #EXs : In BSM (as uracil)\n",
    "#     ['XAN','cpd00309'], #EXs : Xanthine\n",
    "#     ['Xanthosine','cpd01217'], #EXs : \n",
    "#     ['dATP','cpd00115'], #EXs : \n",
    "#     ['dGTP','cpd00241'], #EXs : \n",
    "#     ['dTMP','cpd00298'] #EXs : \n",
    "]\n",
    "\n",
    "# Check to see if these metabolites are used in pathways? Should I add some of these to media? \n",
    "# Yes for ATP, and GTP. (TTP, CTP as well?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['H+', 'cpd00067'],\n",
       " ['H2O', 'cpd00001'],\n",
       " ['CO2', 'cpd00011'],\n",
       " ['O2', 'cpd00007'],\n",
       " ['N2', 'cpd00528'],\n",
       " ['K+', 'cpd00205'],\n",
       " ['Na+', 'cpd00971'],\n",
       " ['Mg', 'cpd00254'],\n",
       " ['Mn2+', 'cpd00030'],\n",
       " ['Fe2+', 'cpd10515'],\n",
       " ['Ca2+', 'cpd00063'],\n",
       " ['Vitamin B12r', 'cpd00423'],\n",
       " ['Cobinamide', 'cpd03422'],\n",
       " ['BIOT', 'cpd00104'],\n",
       " ['PAN', 'cpd00644'],\n",
       " ['Folate', 'cpd00393'],\n",
       " ['Niacin', 'cpd00218'],\n",
       " ['Pyridoxal', 'cpd00215'],\n",
       " ['Riboflavin', 'cpd00220'],\n",
       " ['thiamin', 'cpd00305'],\n",
       " ['Thioglycolate', 'cpd01415'],\n",
       " ['Acetate', 'cpd00029'],\n",
       " ['Citrate', 'cpd00137'],\n",
       " ['ABEE', 'cpd00443'],\n",
       " ['D-Glucose', 'cpd00027'],\n",
       " ['NH3', 'cpd00013'],\n",
       " ['Phosphate', 'cpd00009'],\n",
       " ['Sulfate', 'cpd00048'],\n",
       " ['Adenosine', 'cpd00182'],\n",
       " ['Cytosine', 'cpd00307'],\n",
       " ['Guanosine', 'cpd00311'],\n",
       " ['Thymidine', 'cpd00184'],\n",
       " ['Uridine', 'cpd00249']]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bsm + M9_sources + rna_bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Carbon Sources from all Exchanges and additional interesting sources\n",
    "[\n",
    "    ['4-Hydroxybenzoate','cpd00136'], #EXs : found in coconuts\n",
    "    ['2-keto-3-deoxygluconate','cpd00176'], #EXs : degraded pectin product\n",
    "    ['Amylotriose','cpd01262'], #EXs : \n",
    "    ['CELB','cpd00158'], #EXs : Cellobiose\n",
    "    ['D-Fructose','cpd00082'], #EXs : \n",
    "    ['D-Glucose','cpd00027'], #EXs : \n",
    "    ['D-Mannitol','cpd00314'], #EXs : sweetener the is poorly absorbed in the gut\n",
    "    ['D-Mannose','cpd00138'], #EXs : related to mucin\n",
    "    ['Ribose','cpd00105'], #EXs : \n",
    "    ['Dextrin','cpd11594'], #EXs : \n",
    "    ['Dulcose','cpd01171'], #EXs : Galactitol\n",
    "    ['GLCN','cpd00222'], #EXs : Gluconate \n",
    "    ['GLUM','cpd00276'], #EXs : Glucosamine\n",
    "    ['Galactose','cpd00108'], #EXs : \n",
    "    ['L-Arabinose','cpd00224'], #EXs : \n",
    "    ['L-Inositol','cpd00121'], #EXs : \n",
    "    ['L-Lactate','cpd00159'], #EXs : \n",
    "    ['L-Malate','cpd00130'], #EXs : \n",
    "    ['Glycerol','cpd00100'], #EXs : \n",
    "    ['LACT','cpd00208'], #EXs : lactose\n",
    "    ['Maltohexaose','cpd01329'], #EXs : \n",
    "    ['Maltose','cpd00179'], #EXs : \n",
    "    ['Melibiose','cpd03198'], #EXs : \n",
    "    ['Palmitate','cpd00214'], #EXs : \n",
    "    ['Propionate','cpd00141'], #EXs : \n",
    "    ['Salicin','cpd01030'], #EXs : \n",
    "    ['Sorbitol','cpd00588'], #EXs : \n",
    "    ['Stachyose','cpd01133'], #EXs : \n",
    "    ['Succinate','cpd00036'], #EXs : \n",
    "    ['Sucrose','cpd00076'], #EXs : \n",
    "    ['TRHL','cpd00794'], #EXs : Trehalose\n",
    "    ['Ursin','cpd03696'], #EXs : Arbutin\n",
    "    ['Xylose','cpd00154'], #EXs : \n",
    "    ['hexadecenoate','cpd15237'] #EXs : \n",
    "]\n",
    "\n",
    "# Nitrogen Sources\n",
    "[\n",
    "#     ['NH3','cpd00013'], #EXs : \n",
    "    ['Allantoin','cpd01092'], #EXs : degradation product of purines\n",
    "    ['BET','cpd00540'], #EXs : Betaine\n",
    "    ['Choline','cpd00098'], #EXs : Found in milk\n",
    "    ['GABA','cpd00281'], #EXs : Could also be a carbon source\n",
    "    ['Nitrate','cpd00209'], #EXs : \n",
    "    ['Nitrite','cpd00075'], #EXs : \n",
    "    ['Spermidine','cpd00264'], #EXs : \n",
    "    ['Urea','cpd00073'], #EXs : \n",
    "    ['crotonobetaine','cpd08305'] #EXs : \n",
    "]\n",
    "\n",
    "# Sulfur Sources\n",
    "[\n",
    "    ['H2S2O3','cpd00268'], #EXs : Thiosulfate\n",
    "    ['Isethionate','cpd03048'], #EXs : C2H5O4S\n",
    "#     ['Sulfate','cpd00048'], #EXs : O4S\n",
    "    ['Sulfite','cpd00081'], #EXs : HO3S\n",
    "    ['Sulfoacetate','cpd09878'], #EXs : C2H2O5S\n",
    "    ['ethanesulfonate','cpd11579'], #EXs : C2H5O3S\n",
    "    ['methanesulfonate','cpd08023'] #EXs : CH3O3S\n",
    "]\n",
    "\n",
    "# Phosphorus Sources\n",
    "[\n",
    "    ['Phosphate','cpd00009'] #EX :\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amino Acid related metabolites\n",
    "aas = [\n",
    "    ['D-Alanine','cpd00117'], #EXs : \n",
    "    ['D-Glutamate','cpd00186'], #EXs : \n",
    "    ['D-Methionine','cpd00637'], #EXs : \n",
    "    ['D-Serine','cpd00550'], #EXs : \n",
    "    ['Glycine','cpd00033'], #EXs : 1\n",
    "    ['L-Alanine','cpd00035'], #EXs : 2\n",
    "    ['L-Arginine','cpd00051'], #EXs : 3\n",
    "    ['L-Asparagine','cpd00132'], #EXs : 4\n",
    "    ['L-Aspartate','cpd00041'], #EXs : 5\n",
    "\n",
    "    ['L-Cysteine','cpd00084'], #EXs : 7\n",
    "    ['L-Glutamate','cpd00023'], #EXs : 8\n",
    "    ['L-Glutamine','cpd00053'], #EXs : 9\n",
    "    ['L-Histidine','cpd00119'], #EXs : 10\n",
    "    ['L-Isoleucine','cpd00322'], #EXs : 11\n",
    "    ['L-Leucine','cpd00107'], #EXs : 12\n",
    "    ['L-Lysine','cpd00039'], #EXs : 13\n",
    "    ['L-Methionine','cpd00060'], #EXs : 14\n",
    "    ['L-Phenylalanine','cpd00066'], #EXs : 15\n",
    "    ['L-Proline','cpd00129'], #EXs : 16\n",
    "    ['L-Serine','cpd00054'], #EXs : 17\n",
    "    ['L-Threonine','cpd00161'], #EXs : 18\n",
    "    ['L-Tryptophan','cpd00065'], #EXs : 19\n",
    "    ['L-Tyrosine','cpd00069'], #EXs : 20\n",
    "    ['L-Valine','cpd00156'] #EXs : 21\n",
    "]\n",
    "# Explore leave one out with core amino acids. \n",
    "\n",
    "# Dimers, and other amino acid related mets\n",
    "aa_related = [\n",
    "    ['2-Oxoglutarate','cpd00024'], #EXs : \n",
    "    ['Ala-Gln','cpd11587'], #EXs : \n",
    "    ['Ala-His','cpd11584'], #EXs : \n",
    "    ['Ala-Leu','cpd11583'], #EXs : \n",
    "    ['ala-L-asp-L','cpd11593'], #EXs : \n",
    "    ['ala-L-glu-L','cpd11586'], #EXs : \n",
    "    ['ala-L-Thr-L','cpd11582'], #EXs : \n",
    "    ['Aminoethanol','cpd00162'], #EXs : Ethanolamine\n",
    "    ['Carnitine','cpd00266'], #EXs : \n",
    "    ['Chorismate','cpd00216'], #EXs : \n",
    "    ['L-Cysteate','cpd00395'], #EXs : \n",
    "    ['Cys-Gly','cpd01017'], #EXs : \n",
    "    ['Gly-Cys','cpd15603'], #EXs : \n",
    "    ['Gly-Gln','cpd11580'], #EXs : \n",
    "    ['Gly-Leu','cpd15604'], #EXs : \n",
    "    ['Gly-Met','cpd11591'], #EXs : \n",
    "    ['Gly-Phe','cpd15605'], #EXs : \n",
    "    ['Gly-Tyr','cpd15606'], #EXs : \n",
    "    ['gly-asn-L','cpd11581'], #EXs : \n",
    "    ['gly-asp-L','cpd11589'], #EXs : \n",
    "    ['gly-glu-L','cpd11592'], #EXs : \n",
    "    ['gly-pro-L','cpd11588'], #EXs : \n",
    "    ['L-Methionine S-oxide','cpd01914'], #EXs :\n",
    "    ['L-alanylglycine','cpd11585'], #EXs : \n",
    "    ['L-methionine R-oxide','cpd11576'], #EXs : \n",
    "    ['met-L-ala-L','cpd11590'], #EXs :\n",
    "    ['S-Adenosyl-L-methionine','cpd00017'], #EXs : \n",
    "    ['S-Methyl-L-methionine','cpd02027'], #EXs : \n",
    "    ['S-Ribosylhomocysteine','cpd02227'], #EXs : \n",
    "    ['N-Acetyl-D-glucosamine','cpd00122'], #EXs : \n",
    "    ['N-Acetyl-D-mannosamine','cpd00492'], #EXs : \n",
    "    ['Ornithine','cpd00064'], #EXs : \n",
    "    ['Putrescine','cpd00118'], #EXs : \n",
    "    ['Taurine','cpd00210'], #EXs : \n",
    "    ['meso-2,6-Diaminopimelate','cpd00516'] #EXs : related to lysine\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Production\n",
    "# H2O2 -- cpd00025\n",
    "# Acetate -- cpd00029\n",
    "# Butyrate -- cpd00211\n",
    "# isobutyrate -- cpd01711\n",
    "# GABA -- cpd00281\n",
    "# ethanol -- cpd00363\n",
    "# Propionate -- cpd00141\n",
    "# formate -- cpd00047\n",
    "# Valerate -- cpd00597\n",
    "# Isovaleric acid -- cpd05178 (wrong eqn)\n",
    "# sulforaphane -- \n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5541232/\n",
    "# thiamin -- cpd00305\n",
    "# Pyridoxal phosphate (B6) -- cpd00016\n",
    "# BIOT (biotin, B7) -- cpd00104\n",
    "# (CH3)3NO (TMAO) -- cpd00811\n",
    "# Indole-3-(carb)aldehyde -- cpd05401\n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4337795/\n",
    "# Acetaldehyde -- cpd00071\n",
    "# Deoxycholate -- cpd02733\n",
    "# Chorismate -- cpd00216\n",
    "# Hexanoate -- cpd01113\n",
    "# \n",
    "\n",
    "# Consumption\n",
    "# Galactose -- cpd00108; cpd01112\n",
    "# L-galactose -- cpd01257\n",
    "# lactose -- cpd00208\n",
    "# beta-lactose -- cpd01354\n",
    "# sucrose -- cpd00076\n",
    "# trehalose (TRHL) -- cpd00794\n",
    "# maltose -- cpd00179\n",
    "# D-Mannose -- cpd00138\n",
    "# D-Fructose -- cpd00082\n",
    "# Inulin -- cpd27312\n",
    "# ethanol -- cpd00363\n",
    "# Carnitine -- cpd00266\n",
    "# Citrate -- cpd00137\n",
    "# GLUM (D-glucosamine) -- cpd00276\n",
    "# \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting Loop\n",
      "0...gapfilling...optimizing...Average Likelihood of: cpd00156 is 0.45168461047534947\n",
      "Time to complete:1.6734645168mins\n",
      "\n",
      "\n",
      "[[<Reaction rxn00191_c at 0x7eff879c2890>]]\n",
      "[0.0, 0.7297578149014129, 0.7321378826037916, 0.5191220214505604, 0.6537482407428824, 0.7393325875973646, 0.24052218400619058, 0.7393325875973646, 0.025276721182982728, 0.5916669155435752, 0.2510839990368638, 0.6189340413235531, 0.756340090627988, 0.7336237270826534, 0.5819488983073209, 0.08116948790590572, 0.13632578864588127, 0.0]\n"
     ]
    }
   ],
   "source": [
    "# Using pFBA\n",
    "\n",
    "t = time.time()\n",
    "counter = 0\n",
    "\n",
    "universal = cobra.io.load_json_model(\"../Data/GramPosUni.json\")\n",
    "genome_id = '220668.9'\n",
    "model = cobra.io.read_sbml_model('../gap_models/'+ genome_id +'.xml')\n",
    "likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "\n",
    "# Ensure free water exhange\n",
    "model.reactions.get_by_id('rxn05319_c').name = \"Water transport\"\n",
    "model.reactions.get_by_id('rxn05319_c').bounds = (-1000., 1000.)\n",
    "\n",
    "# Create specific Media List\n",
    "media_list = bsm + M9_sources\n",
    "set_media(model, media_list, universal, verbose=False)\n",
    "\n",
    "# Run through each amino acid to check for production\n",
    "aa_like = {}\n",
    "sys.stdout.write('Starting Loop')\n",
    "# for aa_list in aas[0:1]:\n",
    "\n",
    "sys.stdout.write('\\n'+ str(counter))\n",
    "# aa = aa_list[1]+'_c'\n",
    "aa = 'cpd00117_c'\n",
    "# Add Demand Reaction for metabolite\n",
    "metabolite = model.metabolites.get_by_id(aa)\n",
    "demand = model.add_boundary(metabolite, type='demand')\n",
    "model.objective = demand\n",
    "# Gapfill\n",
    "sys.stdout.write('...gapfilling...')\n",
    "gaps_to_fill = gapfill(model, universal, demand_reactions=False) # Update to probannopy gapfill function; use Gurobi\n",
    "# Fill the gaps\n",
    "rxns_to_add = []\n",
    "for gap in gaps_to_fill:\n",
    "    model.add_reactions(gap)\n",
    "# Optimize with full pathway\n",
    "sys.stdout.write('optimizing...')\n",
    "solution = pfba(model, objective = demand)\n",
    "# Find reactions that carry flux\n",
    "df = solution.fluxes.to_frame()\n",
    "active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "# Add demand reactions for any metabolites produced, but not consumed\n",
    "\n",
    "# Acquire likelihood scores for reactions that carry flux\n",
    "like_list = []\n",
    "rxns_w_likes = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        try:\n",
    "            like_list.append(likelihoods[rxn])\n",
    "            rxns_w_likes.append(rxn)\n",
    "        except:\n",
    "            pass\n",
    "avg_like = np.mean(like_list)\n",
    "sys.stdout.write('Average Likelihood of: ' + aa_list[1] + ' is ' + str(avg_like))\n",
    "aa_like[aa_list[1]] = avg_like\n",
    "# Remove demand and filled gaps\n",
    "model.remove_reactions([demand])\n",
    "model.remove_reactions(gaps_to_fill[0]) # This might not work when multiple reactions are added... \n",
    "\n",
    "    ## Remove demands added to remove extra metabolites produced\n",
    "    \n",
    "counter += 1\n",
    "\n",
    "elapsed = time.time() - t\n",
    "print(\"\\nTime to complete:\" + str(elapsed/60) + \"mins\")\n",
    "print('\\n')\n",
    "print(gaps_to_fill)\n",
    "print(rxns_w_likes)\n",
    "print(like_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[<Reaction rxn00191_c at 0x7eff879c2890>]]\n",
      "['rxn00062_c', 'rxn00545_c', 'rxn00148_c', 'rxn00781_c', 'rxn05466_c', 'rxn01280_c', 'rxn00747_c', 'rxn01281_c', 'rxn00786_c', 'rxn01100_c', 'rxn01106_c', 'rxn10042_c', 'rxn00283_c', 'rxn00184_c', 'rxn00459_c', 'rxn05226_c', 'rxn00558_c', u'rxn00191_c']\n",
      "[0.0, 0.7297578149014129, 0.7321378826037916, 0.5191220214505604, 0.6537482407428824, 0.7393325875973646, 0.24052218400619058, 0.7393325875973646, 0.025276721182982728, 0.5916669155435752, 0.2510839990368638, 0.6189340413235531, 0.756340090627988, 0.7336237270826534, 0.5819488983073209, 0.08116948790590572, 0.13632578864588127, 0.0]\n"
     ]
    }
   ],
   "source": [
    "print(gaps_to_fill)\n",
    "print(rxns_w_likes)\n",
    "print(like_list)\n",
    "# print(likelihoods['rxn00904_c'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "like_list = []\n",
    "rxns_w_likes = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        try:\n",
    "            like_list.append(likelihoods[rxn])\n",
    "            rxns_w_likes.append(rxn)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rxn00062_c\n",
      "rxn00545_c\n",
      "rxn00148_c\n",
      "rxn00781_c\n",
      "rxn05319_c\n",
      "rxn05466_c\n",
      "rxn01280_c\n",
      "rxn00747_c\n",
      "rxn01281_c\n",
      "rxn00786_c\n",
      "rxn01100_c\n",
      "rxn01106_c\n",
      "rxn10042_c\n",
      "rxn00283_c\n",
      "rxn00184_c\n",
      "rxn00459_c\n",
      "rxn05226_c\n",
      "rxn00558_c\n",
      "rxn00191_c\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cpd00001_c',\n",
       " 'cpd00002_c',\n",
       " 'cpd00003_c',\n",
       " 'cpd00004_c',\n",
       " 'cpd00005_c',\n",
       " 'cpd00006_c',\n",
       " 'cpd00008_c',\n",
       " 'cpd00009_c',\n",
       " 'cpd00013_c',\n",
       " 'cpd00020_c',\n",
       " 'cpd00023_c',\n",
       " 'cpd00024_c',\n",
       " 'cpd00035_c',\n",
       " 'cpd00061_c',\n",
       " 'cpd00067_c',\n",
       " 'cpd00072_c',\n",
       " 'cpd00079_c',\n",
       " 'cpd00095_c',\n",
       " 'cpd00102_c',\n",
       " 'cpd00169_c',\n",
       " 'cpd00203_c',\n",
       " 'cpd00223_c',\n",
       " 'cpd00290_c',\n",
       " 'cpd00482_c',\n",
       " 'cpd00843_c'}"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_mets = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        print(rxn)\n",
    "        rxn_obj = model.reactions.get_by_id(rxn)\n",
    "        for met in rxn_obj.metabolites:\n",
    "            if (met.id.endswith('_c')) & (met.id != aa):\n",
    "                all_mets.append(met.id)\n",
    "all_mets = set(all_mets)\n",
    "all_mets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rxn00062_c</th>\n",
       "      <td>750.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00545_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00148_c</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00781_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn05319_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn05466_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01280_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00747_c</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01281_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00786_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01100_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01106_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn10042_c</th>\n",
       "      <td>-250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00283_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00184_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00459_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn05226_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00558_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cpd00067_e</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cpd00001_e</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cpd00013_e</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cpd00027_e</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DM_cpd00117_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00191_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               fluxes\n",
       "rxn00062_c      750.0\n",
       "rxn00545_c      500.0\n",
       "rxn00148_c     -500.0\n",
       "rxn00781_c     1000.0\n",
       "rxn05319_c     1000.0\n",
       "rxn05466_c     1000.0\n",
       "rxn01280_c    -1000.0\n",
       "rxn00747_c     -500.0\n",
       "rxn01281_c     1000.0\n",
       "rxn00786_c      500.0\n",
       "rxn01100_c    -1000.0\n",
       "rxn01106_c    -1000.0\n",
       "rxn10042_c     -250.0\n",
       "rxn00283_c     1000.0\n",
       "rxn00184_c    -1000.0\n",
       "rxn00459_c     1000.0\n",
       "rxn05226_c      500.0\n",
       "rxn00558_c      500.0\n",
       "EX_cpd00067_e  1000.0\n",
       "EX_cpd00001_e  1000.0\n",
       "EX_cpd00013_e -1000.0\n",
       "EX_cpd00027_e  -500.0\n",
       "DM_cpd00117_c  1000.0\n",
       "rxn00191_c    -1000.0"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "active"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "boundary DM_cpd00117_c already exists",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-178-e277db21fe9e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmetabolite\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetabolites\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maa\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdemand\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_boundary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetabolite\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'demand'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobjective\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdemand\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/tjmoutinho/.local/lib/python2.7/site-packages/cobra/core/model.pyc\u001b[0m in \u001b[0;36madd_boundary\u001b[0;34m(self, metabolite, type, reaction_id, lb, ub)\u001b[0m\n\u001b[1;32m    505\u001b[0m             \u001b[0mreaction_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"{}_{}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprefix\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmetabolite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    506\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mreaction_id\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreactions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 507\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'boundary %s already exists'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mreaction_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    508\u001b[0m         \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"{} {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetabolite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    509\u001b[0m         rxn = Reaction(id=reaction_id, name=name, lower_bound=lb,\n",
      "\u001b[0;31mValueError\u001b[0m: boundary DM_cpd00117_c already exists"
     ]
    }
   ],
   "source": [
    "metabolite = model.metabolites.get_by_id(aa)\n",
    "demand = model.add_boundary(metabolite, type='demand')\n",
    "model.objective = demand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DM_cpd00117_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-Alanine demand</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07effa9027890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00117_c --> </p>\n",
       "                    <p style='text-align:right'>D-Alanine --> </p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DM_cpd00117_c at 0x7effa9027890>"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('DM_cpd00117_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in all_mets:\n",
    "    metabolite = model.metabolites.get_by_id(met)\n",
    "    model.add_boundary(metabolite, type='demand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "140"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.boundary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average Likelihood of: cpd00117_c is 0.32383926223275517"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rxn00148_c</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01102_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00781_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn10481_c</th>\n",
       "      <td>250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn05466_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00747_c</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01100_c</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00627_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01870_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01106_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn03366_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00283_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00184_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn05528_c</th>\n",
       "      <td>250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00223_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00459_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn03365_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn05226_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00216_c</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn01286_c</th>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cpd00013_e</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cpd00027_e</th>\n",
       "      <td>-500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rxn00191_c</th>\n",
       "      <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DM_cpd00117_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DM_cpd00001_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DM_cpd00067_c</th>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               fluxes\n",
       "rxn00148_c     -500.0\n",
       "rxn01102_c      500.0\n",
       "rxn00781_c      500.0\n",
       "rxn10481_c      250.0\n",
       "rxn05466_c     1000.0\n",
       "rxn00747_c     -500.0\n",
       "rxn01100_c     -500.0\n",
       "rxn00627_c      500.0\n",
       "rxn01870_c      500.0\n",
       "rxn01106_c    -1000.0\n",
       "rxn03366_c     1000.0\n",
       "rxn00283_c     1000.0\n",
       "rxn00184_c    -1000.0\n",
       "rxn05528_c      250.0\n",
       "rxn00223_c      500.0\n",
       "rxn00459_c     1000.0\n",
       "rxn03365_c    -1000.0\n",
       "rxn05226_c      500.0\n",
       "rxn00216_c     -500.0\n",
       "rxn01286_c      500.0\n",
       "EX_cpd00013_e -1000.0\n",
       "EX_cpd00027_e  -500.0\n",
       "rxn00191_c    -1000.0\n",
       "DM_cpd00117_c  1000.0\n",
       "DM_cpd00001_c  1000.0\n",
       "DM_cpd00067_c  1000.0"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution = pfba(model, objective = demand)\n",
    "# Find reactions that carry flux\n",
    "df = solution.fluxes.to_frame()\n",
    "active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "# Add demand reactions for any metabolites produced, but not consumed\n",
    "\n",
    "# Acquire likelihood scores for reactions that carry flux\n",
    "like_list = []\n",
    "rxns_w_likes = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        try:\n",
    "            like_list.append(likelihoods[rxn])\n",
    "            rxns_w_likes.append(rxn)\n",
    "        except:\n",
    "            pass\n",
    "avg_like = np.mean(like_list)\n",
    "sys.stdout.write('Average Likelihood of: ' + aa + ' is ' + str(avg_like))\n",
    "aa_like[aa_list[1]] = avg_like\n",
    "\n",
    "active"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['rxn00148_c', 'rxn01102_c', 'rxn00781_c', 'rxn10481_c', 'rxn05466_c', 'rxn00747_c', 'rxn01100_c', 'rxn00627_c', 'rxn01870_c', 'rxn01106_c', 'rxn03366_c', 'rxn00283_c', 'rxn00184_c', 'rxn05528_c', 'rxn00223_c', 'rxn00459_c', 'rxn03365_c', 'rxn05226_c', 'rxn00216_c', 'rxn01286_c', u'rxn00191_c']\n",
      "[0.7321378826037916, 0.7573252964069241, 0.5191220214505604, 0.0016726659094700488, 0.6537482407428824, 0.24052218400619058, 0.5916669155435752, 0.0, 0.025276721182982728, 0.2510839990368638, 0.0, 0.756340090627988, 0.7336237270826534, 0.0016726659094700488, 0.0, 0.5819488983073209, 0.0, 0.08116948790590572, 0.4759403048212528, 0.39737340535002535, 0.0]\n"
     ]
    }
   ],
   "source": [
    "print(rxns_w_likes)\n",
    "print(like_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DM_cpd00067_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>H+ demand</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07eff84d3bcd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>cpd00067_c --> </p>\n",
       "                    <p style='text-align:right'>H+ --> </p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DM_cpd00067_c at 0x7eff84d3bcd0>"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('DM_cpd00067_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Reaction rxn13783_c at 0x7eff899e4b10>,\n",
       " <Reaction rxn13784_c at 0x7eff865b1510>,\n",
       " <Reaction rxn13782_c at 0x7eff8670d590>,\n",
       " <Reaction EX_cpd00067_e at 0x7eff8e154d10>,\n",
       " <Reaction EX_cpd11576_e at 0x7eff8e154c90>,\n",
       " <Reaction EX_cpd04097_e at 0x7eff8e154bd0>,\n",
       " <Reaction EX_cpd00033_e at 0x7eff8e154ad0>,\n",
       " <Reaction EX_cpd11585_e at 0x7eff8e154b10>,\n",
       " <Reaction EX_cpd11582_e at 0x7eff8e154950>,\n",
       " <Reaction EX_cpd11586_e at 0x7eff8e154990>,\n",
       " <Reaction EX_cpd01030_e at 0x7eff8e154850>,\n",
       " <Reaction EX_cpd01012_e at 0x7eff8e1546d0>,\n",
       " <Reaction EX_cpd00276_e at 0x7eff8e154750>,\n",
       " <Reaction EX_cpd00082_e at 0x7eff8e1545d0>,\n",
       " <Reaction EX_cpd00080_e at 0x7eff8e154590>,\n",
       " <Reaction EX_cpd00001_e at 0x7eff8e154510>,\n",
       " <Reaction EX_cpd00023_e at 0x7eff8e154490>,\n",
       " <Reaction EX_cpd00971_e at 0x7eff8e154350>,\n",
       " <Reaction EX_cpd00058_e at 0x7eff8e1541d0>,\n",
       " <Reaction EX_cpd00224_e at 0x7eff8e154250>,\n",
       " <Reaction EX_cpd00314_e at 0x7eff8e154110>,\n",
       " <Reaction EX_cpd01017_e at 0x7eff8e1540d0>,\n",
       " <Reaction EX_cpd00222_e at 0x7eff8e154150>,\n",
       " <Reaction EX_cpd00013_e at 0x7eff8e141f10>,\n",
       " <Reaction EX_cpd10515_e at 0x7eff8e141ed0>,\n",
       " <Reaction EX_cpd00179_e at 0x7eff8e141d90>,\n",
       " <Reaction EX_cpd00100_e at 0x7eff8e141e50>,\n",
       " <Reaction EX_cpd11591_e at 0x7eff8e141c10>,\n",
       " <Reaction EX_cpd00009_e at 0x7eff8e141c90>,\n",
       " <Reaction EX_cpd00060_e at 0x7eff8e141b90>,\n",
       " <Reaction EX_cpd00681_e at 0x7eff8e141b50>,\n",
       " <Reaction EX_cpd00034_e at 0x7eff8e141a50>,\n",
       " <Reaction EX_cpd00098_e at 0x7eff8e141990>,\n",
       " <Reaction EX_cpd11593_e at 0x7eff8e1417d0>,\n",
       " <Reaction EX_cpd00027_e at 0x7eff8e141850>,\n",
       " <Reaction EX_cpd00076_e at 0x7eff8e1415d0>,\n",
       " <Reaction EX_cpd01171_e at 0x7eff8e141650>,\n",
       " <Reaction EX_cpd15605_e at 0x7eff8e141690>,\n",
       " <Reaction EX_cpd00550_e at 0x7eff8e141550>,\n",
       " <Reaction EX_cpd00054_e at 0x7eff8e141490>,\n",
       " <Reaction EX_cpd00154_e at 0x7eff8e141410>,\n",
       " <Reaction EX_cpd03724_e at 0x7eff8e141450>,\n",
       " <Reaction EX_cpd00637_e at 0x7eff8e1412d0>,\n",
       " <Reaction EX_cpd00122_e at 0x7eff8e141110>,\n",
       " <Reaction EX_cpd08306_e at 0x7eff8e141050>,\n",
       " <Reaction EX_cpd15603_e at 0x7eff8e1410d0>,\n",
       " <Reaction EX_cpd03696_e at 0x7eff8e130e90>,\n",
       " <Reaction EX_cpd00309_e at 0x7eff8e130f10>,\n",
       " <Reaction EX_cpd11580_e at 0x7eff8e130e10>,\n",
       " <Reaction EX_cpd01092_e at 0x7eff8e130dd0>,\n",
       " <Reaction EX_cpd00220_e at 0x7eff8e130c90>,\n",
       " <Reaction EX_cpd00588_e at 0x7eff8e130cd0>,\n",
       " <Reaction EX_cpd00117_e at 0x7eff8e130c50>,\n",
       " <Reaction EX_cpd00108_e at 0x7eff8e130ad0>,\n",
       " <Reaction EX_cpd00130_e at 0x7eff8e130a50>,\n",
       " <Reaction EX_cpd00024_e at 0x7eff8e130950>,\n",
       " <Reaction EX_cpd00307_e at 0x7eff8e1309d0>,\n",
       " <Reaction EX_cpd00254_e at 0x7eff8e1307d0>,\n",
       " <Reaction EX_cpd00138_e at 0x7eff8e130810>,\n",
       " <Reaction EX_cpd00305_e at 0x7eff8e130690>,\n",
       " <Reaction EX_cpd00159_e at 0x7eff8e130710>,\n",
       " <Reaction EX_cpd00039_e at 0x7eff8e1304d0>,\n",
       " <Reaction EX_cpd00540_e at 0x7eff8e130550>,\n",
       " <Reaction EX_cpd15606_e at 0x7eff8e130590>,\n",
       " <Reaction EX_cpd10516_e at 0x7eff8e130410>,\n",
       " <Reaction EX_cpd00030_e at 0x7eff8e130310>,\n",
       " <Reaction EX_cpd00149_e at 0x7eff8e1302d0>,\n",
       " <Reaction EX_cpd11597_e at 0x7eff8e130250>,\n",
       " <Reaction EX_cpd01914_e at 0x7eff8e130090>,\n",
       " <Reaction EX_cpd00531_e at 0x7eff8e130110>,\n",
       " <Reaction EX_cpd11583_e at 0x7eff8e130190>,\n",
       " <Reaction EX_cpd11589_e at 0x7eff8e19ddd0>,\n",
       " <Reaction EX_cpd00794_e at 0x7eff8e19de50>,\n",
       " <Reaction EX_cpd00105_e at 0x7eff8e19ded0>,\n",
       " <Reaction EX_cpd00226_e at 0x7eff8e19df50>,\n",
       " <Reaction EX_cpd08305_e at 0x7eff8e19dcd0>,\n",
       " <Reaction EX_cpd11588_e at 0x7eff8e19dd50>,\n",
       " <Reaction EX_cpd15604_e at 0x7eff8e19dc10>,\n",
       " <Reaction EX_cpd00104_e at 0x7eff8e19db10>,\n",
       " <Reaction EX_cpd00007_e at 0x7eff8e19dad0>,\n",
       " <Reaction EX_cpd11592_e at 0x7eff8e19da10>,\n",
       " <Reaction EX_cpd11587_e at 0x7eff8e19d910>,\n",
       " <Reaction EX_cpd00092_e at 0x7eff8e19d990>,\n",
       " <Reaction EX_cpd00053_e at 0x7eff8e19d790>,\n",
       " <Reaction EX_cpd11590_e at 0x7eff8e19d850>,\n",
       " <Reaction EX_cpd11581_e at 0x7eff8e19d690>,\n",
       " <Reaction EX_cpd00205_e at 0x7eff8e19d5d0>,\n",
       " <Reaction EX_cpd11584_e at 0x7eff8e19d550>,\n",
       " <Reaction EX_cpd00158_e at 0x7eff8e19d490>,\n",
       " <Reaction EX_cpd00121_e at 0x7eff8e19d3d0>,\n",
       " <Reaction EX_cpd00011_e at 0x7eff8e19d2d0>,\n",
       " <Reaction EX_cpd00355_e at 0x7eff8e19d310>,\n",
       " <Reaction EX_cpd11606_e at 0x7eff8e19d210>,\n",
       " <Reaction EX_cpd00393_e at 0x7eff8e19d150>,\n",
       " <Reaction EX_cpd00322_e at 0x7eff8e19d090>,\n",
       " <Reaction EX_cpd00099_e at 0x7eff8e18bfd0>,\n",
       " <Reaction EX_cpd00644_e at 0x7eff8e18be10>,\n",
       " <Reaction EX_cpd01188_e at 0x7eff8e18bf10>,\n",
       " <Reaction EX_cpd00028_e at 0x7eff8e18bed0>,\n",
       " <Reaction EX_cpd00063_e at 0x7eff8e18bcd0>,\n",
       " <Reaction EX_cpd00048_e at 0x7eff8e18bc50>,\n",
       " <Reaction EX_cpd00156_e at 0x7eff8e18bb50>,\n",
       " <Reaction EX_cpd00264_e at 0x7eff8e18bb10>,\n",
       " <Reaction EX_cpd01080_e at 0x7eff8e18bb90>,\n",
       " <Reaction SK_cpd11416_c at 0x7eff8e18bad0>,\n",
       " <Reaction EX_cpd00528_e at 0x7eff823ea110>,\n",
       " <Reaction EX_cpd00423_e at 0x7effa90ba5d0>,\n",
       " <Reaction EX_cpd03422_e at 0x7effa9075990>,\n",
       " <Reaction EX_cpd00218_e at 0x7eff7c299610>,\n",
       " <Reaction EX_cpd00215_e at 0x7effabd8a710>,\n",
       " <Reaction EX_cpd01415_e at 0x7effabd8ae10>,\n",
       " <Reaction EX_cpd00029_e at 0x7effabd8ab10>,\n",
       " <Reaction EX_cpd00137_e at 0x7effabd8a310>,\n",
       " <Reaction EX_cpd00443_e at 0x7effabd8a750>,\n",
       " <Reaction DM_cpd00117_c at 0x7effa9027890>,\n",
       " <Reaction DM_cpd00290_c at 0x7eff7daba550>,\n",
       " <Reaction DM_cpd00013_c at 0x7effa44bee50>,\n",
       " <Reaction DM_cpd00095_c at 0x7eff7dacda50>,\n",
       " <Reaction DM_cpd00482_c at 0x7eff7dacd9d0>,\n",
       " <Reaction DM_cpd00079_c at 0x7eff7dacd150>,\n",
       " <Reaction DM_cpd00023_c at 0x7eff7dacd3d0>,\n",
       " <Reaction DM_cpd00203_c at 0x7eff7dabaad0>,\n",
       " <Reaction DM_cpd00102_c at 0x7eff7daba590>,\n",
       " <Reaction DM_cpd00009_c at 0x7eff7daba450>,\n",
       " <Reaction DM_cpd00006_c at 0x7eff7daba710>,\n",
       " <Reaction DM_cpd00001_c at 0x7eff7d9a7dd0>,\n",
       " <Reaction DM_cpd00008_c at 0x7eff7dacd790>,\n",
       " <Reaction DM_cpd00072_c at 0x7eff7dacdd10>,\n",
       " <Reaction DM_cpd00024_c at 0x7eff84d3ba50>,\n",
       " <Reaction DM_cpd00004_c at 0x7eff84d3b550>,\n",
       " <Reaction DM_cpd00169_c at 0x7eff7dade7d0>,\n",
       " <Reaction DM_cpd00843_c at 0x7eff7daded50>,\n",
       " <Reaction DM_cpd00003_c at 0x7eff7dade390>,\n",
       " <Reaction DM_cpd00020_c at 0x7eff7d995b10>,\n",
       " <Reaction DM_cpd00005_c at 0x7eff7d9957d0>,\n",
       " <Reaction DM_cpd00035_c at 0x7eff7d995a90>,\n",
       " <Reaction DM_cpd00002_c at 0x7eff7d995c10>,\n",
       " <Reaction DM_cpd00067_c at 0x7eff84d3bcd0>,\n",
       " <Reaction DM_cpd00223_c at 0x7eff7b6d45d0>,\n",
       " <Reaction DM_cpd00061_c at 0x7eff7b6d4d10>]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.boundary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in all_mets:\n",
    "    try:\n",
    "        model.remove_reactions([model.reactions.get_by_id('DM_'+met)])\n",
    "    except:\n",
    "        print('skipped')\n",
    "        print(met)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "115"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.boundary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.remove_reactions([model.reactions.get_by_id('DM_cpd00102_c')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'DM_cpd00102_c'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-168-e8f920eb6434>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m[\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreactions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'DM_cpd00102_c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/home/tjmoutinho/.local/lib/python2.7/site-packages/cobra/core/dictlist.pyc\u001b[0m in \u001b[0;36mget_by_id\u001b[0;34m(self, id)\u001b[0m\n\u001b[1;32m     56\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_by_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     57\u001b[0m         \u001b[0;34m\"\"\"return the element with a matching id\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 58\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mlist_attr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattribute\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'DM_cpd00102_c'"
     ]
    }
   ],
   "source": [
    "[model.reactions.get_by_id('DM_cpd00102_c')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "like_list = []\n",
    "rxns_w_likes = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        try:\n",
    "            like_list.append(likelihoods[rxn])\n",
    "            rxns_w_likes.append(rxn)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.00167266590947\n"
     ]
    }
   ],
   "source": [
    "print(likelihoods['rxn10481_c'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.remove_reactions(gaps_to_fill[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model.solver = 'gurobi'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pFBA with new media components (RNA bases + thymidine...)\n",
    "\n",
    "t = time.time()\n",
    "counter = 0\n",
    "\n",
    "universal = cobra.io.load_json_model(\"../Data/GramPosUni.json\")\n",
    "genome_id = '220668.9'\n",
    "model = cobra.io.read_sbml_model('../gap_models/'+ genome_id +'.xml')\n",
    "likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "\n",
    "# Ensure free water exhange\n",
    "model.reactions.get_by_id('rxn05319_c').name = \"Water transport\"\n",
    "model.reactions.get_by_id('rxn05319_c').bounds = (-1000., 1000.)\n",
    "\n",
    "# Create specific Media List\n",
    "media_list = bsm + M9_sources + rna_bases\n",
    "set_media(model, media_list, universal, verbose=False)\n",
    "\n",
    "# Run through each amino acid to check for production\n",
    "aa_like = {}\n",
    "sys.stdout.write('Starting Loop')\n",
    "# for aa_list in aas[0:1]:\n",
    "\n",
    "sys.stdout.write('\\n'+ str(counter))\n",
    "# aa = aa_list[1]+'_c'\n",
    "aa = 'cpd00117_c'\n",
    "# Add Demand Reaction for metabolite\n",
    "metabolite = model.metabolites.get_by_id(aa)\n",
    "demand = model.add_boundary(metabolite, type='demand')\n",
    "model.objective = demand\n",
    "# Gapfill\n",
    "sys.stdout.write('...gapfilling...')\n",
    "gaps_to_fill = gapfill(model, universal, demand_reactions=False) # Update to probannopy gapfill function; use Gurobi\n",
    "# Fill the gaps\n",
    "rxns_to_add = []\n",
    "for gap in gaps_to_fill:\n",
    "    model.add_reactions(gap)\n",
    "# Optimize with full pathway\n",
    "sys.stdout.write('optimizing...')\n",
    "solution = pfba(model, objective = demand)\n",
    "# Find reactions that carry flux\n",
    "df = solution.fluxes.to_frame()\n",
    "active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "# Add demand reactions for all metabolites, not used for optimization, simply there to stop blocked reactions\n",
    "all_mets = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        print(rxn)\n",
    "        rxn_obj = model.reactions.get_by_id(rxn)\n",
    "        for met in rxn_obj.metabolites:\n",
    "            if (met.id.endswith('_c')) & (met.id != aa):\n",
    "                all_mets.append(met.id)\n",
    "all_mets = set(all_mets)\n",
    "all_mets\n",
    "\n",
    "for met in all_mets:\n",
    "    metabolite = model.metabolites.get_by_id(met)\n",
    "    model.add_boundary(metabolite, type='demand')\n",
    "\n",
    "# Acquire likelihood scores for reactions that carry flux\n",
    "like_list = []\n",
    "rxns_w_likes = []\n",
    "for rxn in list(active.index):\n",
    "    if rxn.startswith('rxn'):\n",
    "        try:\n",
    "            like_list.append(likelihoods[rxn])\n",
    "            rxns_w_likes.append(rxn)\n",
    "        except:\n",
    "            pass\n",
    "avg_like = np.mean(like_list)\n",
    "sys.stdout.write('Average Likelihood of: ' + aa_list[1] + ' is ' + str(avg_like))\n",
    "aa_like[aa_list[1]] = avg_like\n",
    "# Remove demand and filled gaps\n",
    "model.remove_reactions([demand])\n",
    "model.remove_reactions(gaps_to_fill[0]) # This might not work when multiple reactions are added... \n",
    "\n",
    "    ## Remove demands added to remove extra metabolites produced\n",
    "    \n",
    "counter += 1\n",
    "\n",
    "elapsed = time.time() - t\n",
    "print(\"\\nTime to complete:\" + str(elapsed/60) + \"mins\")\n",
    "print('\\n')\n",
    "print(gaps_to_fill)\n",
    "print(rxns_w_likes)\n",
    "print(like_list)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
