{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "import cobra\n",
    "import cobra.test\n",
    "# import mackinac\n",
    "import numpy as np\n",
    "import csv\n",
    "import glob\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import time\n",
    "import sys\n",
    "from copy import deepcopy\n",
    "from collections import defaultdict\n",
    "# from cobra.flux_analysis import gapfill\n",
    "from gapfill2 import *\n",
    "from cobra.flux_analysis import pfba\n",
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "from cobra.flux_analysis import sample\n",
    "from cobra.core.solution import get_solution\n",
    "\n",
    "\n",
    "# Set default logger to python logger to avoid warnings given when adding reactions and/or metaboites \n",
    "# because \"cobra.core.model\" doesn't innately have a logger.\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger('logger')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_media(model, media, universal, verbose=False):\n",
    "\n",
    "    # Find and close all exchange reactions in the model\n",
    "    model_rxns = [rxn.id for rxn in model.reactions]\n",
    "    for rxn in model_rxns:\n",
    "        if rxn.startswith('EX_') and rxn.endswith('_e'):\n",
    "            model.reactions.get_by_id(rxn).lower_bound = 0.0\n",
    "\n",
    "    # Check for existence of exchange reactions for the media metabolites in the model\n",
    "    for metabolite in media:\n",
    "        met = metabolite[1]+'_e'\n",
    "        if 'EX_'+met in model_rxns:\n",
    "            model.reactions.get_by_id('EX_'+met).lower_bound = -1000.\n",
    "        else:\n",
    "            # Create exchange reaction and add to model\n",
    "            if verbose:\n",
    "                print(\"added exchange rxn for \" + met)\n",
    "            new_exchange = cobra.Reaction('EX_'+met)\n",
    "            new_exchange.name = met + ' exchange'\n",
    "            met_obj = universal.metabolites.get_by_id(met)\n",
    "            new_exchange.add_metabolites({met_obj:-1})\n",
    "            new_exchange.lower_bound = -1000.\n",
    "            new_exchange.upper_bound = 1000.\n",
    "            model.add_reaction(new_exchange)\n",
    "            model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basal Synthetic Media\n",
    "bsm = [\n",
    "    ['H+','cpd00067'],\n",
    "    ['H2O','cpd00001'],\n",
    "    ['CO2','cpd00011'],\n",
    "    ['O2','cpd00007'],\n",
    "    ['N2','cpd00528'], \n",
    "#     ['H2','cpd11640'], # Only with no O2\n",
    "    \n",
    "    ['K+','cpd00205'],\n",
    "    ['Na+','cpd00971'],\n",
    "    ['Mg','cpd00254'],\n",
    "    ['Mn2+','cpd00030'],\n",
    "    ['Fe2+','cpd10515'], # Iron ion in heme\n",
    "    ['Ca2+','cpd00063'], # Calcium pantothenate;cpd19112\n",
    "    \n",
    "    ['Vitamin B12r','cpd00423'], # C62H91CoN13O14P : cobalamin;cpd03424;cpd00730 : not present in any exchange reactions\n",
    "    ['Cobinamide','cpd03422'], #EXs : related to cobalamin (B12) Added to ensure cells have access to B12\n",
    "    ['BIOT','cpd00104'], # C10H15N2O3S : biotin B7\n",
    "    ['PAN','cpd00644'], # C9H16NO5 : Pantothenate B5\n",
    "    ['Folate','cpd00393'], # C19H17N7O6 : B9\n",
    "    ['Niacin','cpd00218'], # C6H4NO2 : B3\n",
    "    ['Pyridoxal','cpd00215'], # C8H9NO3 : B6\n",
    "    ['Riboflavin','cpd00220'], # C17H19N4O6 : B2\n",
    "    ['thiamin','cpd00305'], # C12H17N4OS : B1\n",
    "    \n",
    "#     ['Phosphate','cpd00009'], # HO4P : In M9 Defaults\n",
    "    \n",
    "    ['Thioglycolate','cpd01415'], # C2H3O2S : not present in any exchange reactions\n",
    "#     ['Sulfate','cpd00048'], # O4S : In M9 Defaults\n",
    "    \n",
    "    ['Acetate','cpd00029'], # C2H3O2 : not present in any exchange reactions\n",
    "    ['Citrate','cpd00137'], # C6H5O7 : Consider removing. \n",
    "#     ['Polysorbate 60','cpd24450'], # C35H68O10 : Almost tween 80 : not present in any reactions\n",
    "#     ['Ethyl acetate','cpd00633'], # C4H8O2 : not present in any exchange reactions, only present in one reaction at all\n",
    "    \n",
    "    ['ABEE','cpd00443'] # C7H6NO2 : aminobenzoate : not present in any exchange reactions\n",
    "]\n",
    "\n",
    "# Potentially add to BSM (from M9 media)\n",
    "M9_ions = [\n",
    "    ['Cl-','cpd00099'],\n",
    "    ['Co2+','cpd00149'],\n",
    "    ['Cu2+','cpd00058'],\n",
    "    ['Fe3','cpd10516'],\n",
    "#     ['Sodium molybdate','cpd11145'], # This doesn't connect to anything\n",
    "    ['Ni2+','cpd00244'],\n",
    "    ['Selenate','cpd03396'],\n",
    "    ['Selenite','cpd03387'],\n",
    "    ['Zn2+','cpd00034']\n",
    "]\n",
    "\n",
    "# Enviromental Metabolites with Exchange reactions\n",
    "[\n",
    "#     ['CO2','cpd00011'], #EXs : \n",
    "#     ['Ca2+','cpd00063'], #EXs : \n",
    "#     ['Cd2+','cpd01012'], #EXs : Removed because toxic\n",
    "#     ['chromate','cpd11595'], #EXs : Removed because toxic\n",
    "#     ['Cl-','cpd00099'], #EXs : \n",
    "#     ['Co2+','cpd00149'], #EXs : In M9\n",
    "#     ['Cu2+','cpd00058'], #EXs : In M9\n",
    "#     ['Fe2+','cpd10515'], #EXs : \n",
    "#     ['H+','cpd00067'], #EXs : \n",
    "#     ['H2','cpd11640'], #EXs : \n",
    "#     ['H2O','cpd00001'], #EXs : \n",
    "#     ['Hg2+','cpd00531'], #EXs : Removed because toxic\n",
    "#     ['K+','cpd00205'], #EXs : \n",
    "#     ['Mg','cpd00254'], #EXs : \n",
    "#     ['Mn2+','cpd00030'], #EXs : \n",
    "#     ['Na+','cpd00971'], #EXs : \n",
    "#     ['Ni2+','cpd00244'], #EXs : In M9\n",
    "#     ['O2','cpd00007'], #EXs : \n",
    "#     ['Pb','cpd04097'], #EXs : Removed because toxic\n",
    "#     ['Zn2+','cpd00034'], #EXs : In M9\n",
    "#     ['fe3','cpd10516'] #EXs : In M9\n",
    "]\n",
    "\n",
    "# M9 Base : https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4932939/\n",
    "# [\n",
    "#     ['Ca2+','cpd00063'],\n",
    "#     ['Cl-','cpd00099'],\n",
    "#     ['CO2','cpd00011'],\n",
    "#     ['Co2+','cpd00149'],\n",
    "#     ['Cu2+','cpd00058'],\n",
    "#     ['Fe2+','cpd10515'],\n",
    "#     ['Fe3','cpd10516'],\n",
    "#     ['H+','cpd00067'],\n",
    "#     ['H2O','cpd00001'],\n",
    "#     ['K+','cpd00205'],\n",
    "#     ['Mg','cpd00254'],\n",
    "#     ['Mn2+','cpd00030'],\n",
    "#     ['Sodium molybdate','cpd11145'],\n",
    "#     ['Na+','cpd00971'],\n",
    "#     ['Ni2+','cpd00244'],\n",
    "#     ['Selenate','cpd03396'],\n",
    "#     ['Selenite','cpd03387'],\n",
    "#     ['Zn2+','cpd00034']\n",
    "# ]\n",
    "\n",
    "# M9 default carbon, nitrogen, phosphorous, and sulfur sources\n",
    "M9_sources = [\n",
    "    ['D-Glucose','cpd00027'],\n",
    "    ['NH3','cpd00013'], # this is actually NH4 : ammonium\n",
    "    ['Phosphate','cpd00009'],\n",
    "    ['Sulfate','cpd00048']\n",
    "]\n",
    "\n",
    "# Vitamins\n",
    "vit_k = [\n",
    "#     ['BIOT','cpd00104'], #EXs : Biotin\n",
    "#     ['Cobinamide','cpd03422'], #EXs : related to cobalamin (B12)\n",
    "#     ['Folate','cpd00393'], #EXs : \n",
    "    ['Menaquinone 7','cpd11606'], #EXs : Vitamine K2 : Add when there is no O2\n",
    "#     ['Niacin','cpd00218'], #EXs : \n",
    "#     ['PAN','cpd00644'], #EXs : Pantothenate\n",
    "#     ['Pyridoxal','cpd00215'], #EXs : \n",
    "#     ['Riboflavin','cpd00220'], #EXs : \n",
    "#     ['Thiamin','cpd00305'] #EXs : \n",
    "]\n",
    "\n",
    "# For aerobic simulations, O2 was added with a lower bound of âˆ’20 and to 0 for anaerobic simulations.\n",
    "\n",
    "# DNA/RNA related metabolites\n",
    "rna_bases = [\n",
    "#     ['35ccmp','cpd00696'], #EXs : \n",
    "#     ['AMP','cpd00018'], #EXs : \n",
    "    ['Adenosine','cpd00182'], #EXs : In BSM (as adenine)\n",
    "#     ['Adenosine 3-5-bisphosphate','cpd00045'], #EXs : \n",
    "    ['Cytosine','cpd00307'], #EXs : \n",
    "#     ['Deoxyadenosine','cpd00438'], #EXs : \n",
    "#     ['Deoxycytidine','cpd00654'], #EXs : \n",
    "#     ['Deoxyguanosine','cpd00277'], #EXs : In BSM\n",
    "#     ['Deoxyinosine','cpd03279'], #EXs : \n",
    "#     ['Deoxyuridine','cpd00412'], #EXs : \n",
    "#     ['GMP','cpd00126'], #EXs : \n",
    "#     ['GTP','cpd00038'], #EXs : \n",
    "    ['Guanosine','cpd00311'], #EXs : In BSM (as Guanine)\n",
    "#     ['Inosine','cpd00246'], #EXs : \n",
    "#     ['HYXN','cpd00226'], #EXs : Hypoxanthine\n",
    "#     ['Nicotinamide ribonucleotide','cpd00355'], #EXs : \n",
    "#     ['TTP','cpd00357'], #EXs : Deoxythymidine triphosphate\n",
    "    ['Thymidine','cpd00184'], #EXs : In BSM\n",
    "#     ['Thyminose','cpd01242'], #EXs : deoxyribose\n",
    "#     ['Uracil','cpd00092'], #EXs : \n",
    "    ['Uridine','cpd00249'], #EXs : In BSM (as uracil)\n",
    "#     ['XAN','cpd00309'], #EXs : Xanthine\n",
    "#     ['Xanthosine','cpd01217'], #EXs : \n",
    "#     ['dATP','cpd00115'], #EXs : \n",
    "#     ['dGTP','cpd00241'], #EXs : \n",
    "#     ['dTMP','cpd00298'] #EXs : \n",
    "]\n",
    "\n",
    "# Check to see if these metabolites are used in pathways? Should I add some of these to media? \n",
    "# Yes for ATP, and GTP. (TTP, CTP as well?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amino Acid related metabolites\n",
    "aas = [\n",
    "    ['D-Alanine','cpd00117'], #EXs : \n",
    "    ['D-Glutamate','cpd00186'], #EXs : \n",
    "    ['D-Methionine','cpd00637'], #EXs : \n",
    "    ['D-Serine','cpd00550'], #EXs : \n",
    "    ['Glycine','cpd00033'], #EXs : 1\n",
    "    ['L-Alanine','cpd00035'], #EXs : 2\n",
    "    ['L-Arginine','cpd00051'], #EXs : 3\n",
    "    ['L-Asparagine','cpd00132'], #EXs : 4\n",
    "    ['L-Aspartate','cpd00041'], #EXs : 5\n",
    "\n",
    "    ['L-Cysteine','cpd00084'], #EXs : 7\n",
    "    ['L-Glutamate','cpd00023'], #EXs : 8\n",
    "    ['L-Glutamine','cpd00053'], #EXs : 9\n",
    "    ['L-Histidine','cpd00119'], #EXs : 10\n",
    "    ['L-Isoleucine','cpd00322'], #EXs : 11\n",
    "    ['L-Leucine','cpd00107'], #EXs : 12\n",
    "    ['L-Lysine','cpd00039'], #EXs : 13\n",
    "    ['L-Methionine','cpd00060'], #EXs : 14\n",
    "    ['L-Phenylalanine','cpd00066'], #EXs : 15\n",
    "    ['L-Proline','cpd00129'], #EXs : 16\n",
    "    ['L-Serine','cpd00054'], #EXs : 17\n",
    "    ['L-Threonine','cpd00161'], #EXs : 18\n",
    "    ['L-Tryptophan','cpd00065'], #EXs : 19\n",
    "    ['L-Tyrosine','cpd00069'], #EXs : 20\n",
    "    ['L-Valine','cpd00156'] #EXs : 21\n",
    "]\n",
    "# Explore leave one out with core amino acids. \n",
    "\n",
    "# Dimers, and other amino acid related mets\n",
    "aa_related = [\n",
    "    ['2-Oxoglutarate','cpd00024'], #EXs : \n",
    "    ['Ala-Gln','cpd11587'], #EXs : \n",
    "    ['Ala-His','cpd11584'], #EXs : \n",
    "    ['Ala-Leu','cpd11583'], #EXs : \n",
    "    ['ala-L-asp-L','cpd11593'], #EXs : \n",
    "    ['ala-L-glu-L','cpd11586'], #EXs : \n",
    "    ['ala-L-Thr-L','cpd11582'], #EXs : \n",
    "    ['Aminoethanol','cpd00162'], #EXs : Ethanolamine\n",
    "    ['Carnitine','cpd00266'], #EXs : \n",
    "    ['Chorismate','cpd00216'], #EXs : \n",
    "    ['L-Cysteate','cpd00395'], #EXs : \n",
    "    ['Cys-Gly','cpd01017'], #EXs : \n",
    "    ['Gly-Cys','cpd15603'], #EXs : \n",
    "    ['Gly-Gln','cpd11580'], #EXs : \n",
    "    ['Gly-Leu','cpd15604'], #EXs : \n",
    "    ['Gly-Met','cpd11591'], #EXs : \n",
    "    ['Gly-Phe','cpd15605'], #EXs : \n",
    "    ['Gly-Tyr','cpd15606'], #EXs : \n",
    "    ['gly-asn-L','cpd11581'], #EXs : \n",
    "    ['gly-asp-L','cpd11589'], #EXs : \n",
    "    ['gly-glu-L','cpd11592'], #EXs : \n",
    "    ['gly-pro-L','cpd11588'], #EXs : \n",
    "    ['L-Methionine S-oxide','cpd01914'], #EXs :\n",
    "    ['L-alanylglycine','cpd11585'], #EXs : \n",
    "    ['L-methionine R-oxide','cpd11576'], #EXs : \n",
    "    ['met-L-ala-L','cpd11590'], #EXs :\n",
    "    ['S-Adenosyl-L-methionine','cpd00017'], #EXs : \n",
    "    ['S-Methyl-L-methionine','cpd02027'], #EXs : \n",
    "    ['S-Ribosylhomocysteine','cpd02227'], #EXs : \n",
    "    ['N-Acetyl-D-glucosamine','cpd00122'], #EXs : \n",
    "    ['N-Acetyl-D-mannosamine','cpd00492'], #EXs : \n",
    "    ['Ornithine','cpd00064'], #EXs : \n",
    "    ['Putrescine','cpd00118'], #EXs : \n",
    "    ['Taurine','cpd00210'], #EXs : \n",
    "    ['meso-2,6-Diaminopimelate','cpd00516'] #EXs : related to lysine\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/python\n",
    "'''\n",
    "Gapfilling function that utilizes pFBA and flux sampling to find most\n",
    "parsimonious additional reactions to achieve minimum flux through the objective\n",
    "Author: Matthew Jenior\n",
    "'''\n",
    "import pandas\n",
    "import math\n",
    "import copy\n",
    "import time\n",
    "import random\n",
    "\n",
    "# Using Cobrapy 0.13.0\n",
    "import cobra\n",
    "import cobra.test\n",
    "from cobra.flux_analysis.sampling import OptGPSampler\n",
    "from cobra.manipulation.delete import *\n",
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "from cobra.medium import find_boundary_types\n",
    "\n",
    "# pFBA gapfiller\n",
    "def pfba_gapfill(model, reaction_bag, likelihoods, obj=None, obj_lb=10., obj_constraint=False,\n",
    "                 iters=1, tasks=None, task_lb=0.05, \n",
    "                 add_exchanges=True, extracellular='e'):\n",
    "    '''\n",
    "    Function that utilizes iterations of pFBA solution with a universal reaction bag \n",
    "    in order to gapfill a model.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    model : cobra.Model\n",
    "        Model to be gapfilled\n",
    "    reaction_bag : cobra.Model\n",
    "        Reaction bag reference to use during gapfilling\n",
    "    obj : string\n",
    "        Reaction ID for objective function in model to be gapfilled.\n",
    "    obj_lb : float\n",
    "        Lower bound for objective function\n",
    "    obj_constraint : bool\n",
    "        Sets objective as contstraint which must be maximized\n",
    "    tasks : list or None\n",
    "        List of reactions IDs (strings) of metabolic tasks \n",
    "        to set a minimum lower bound for\n",
    "    task_lb : float\n",
    "        Lower bound for any metabolic tasks\n",
    "    iters : int\n",
    "        Number of gapfilling rounds. Unique reactions from each round are \n",
    "        saved and the union is added simulatneously to the model\n",
    "    add_exchanges : bool\n",
    "        Identifies extracellular metabolites added during gapfilling that\n",
    "        are not associated with exchange reactions and creates them\n",
    "    extracellular : string\n",
    "        Label for extracellular compartment of model\n",
    "    '''\n",
    "    start_time = time.time()\n",
    "    \n",
    "    # Save some basic network info for downstream membership testing\n",
    "    orig_rxn_ids = set([str(x.id) for x in model.reactions])\n",
    "    orig_cpd_ids = set([str(y.id) for y in model.metabolites])\n",
    "    univ_rxn_ids = set([str(z.id) for z in reaction_bag.reactions])\n",
    "    \n",
    "    # Find overlap in model and reaction bag\n",
    "    overlap_rxn_ids = univ_rxn_ids.intersection(orig_rxn_ids)\n",
    "    \n",
    "    # Get model objective reaction ID\n",
    "    if obj == None:\n",
    "        obj = get_objective(model)\n",
    "    else:\n",
    "        obj = obj\n",
    "    \n",
    "    # Modify universal reaction bag\n",
    "    new_rxn_ids = set()\n",
    "    print('Creating universal model...')\n",
    "    with reaction_bag as universal:\n",
    "\n",
    "        # Remove overlapping reactions from universal bag, and reset objective if needed\n",
    "        for rxn in overlap_rxn_ids: \n",
    "            universal.reactions.get_by_id(rxn).remove_from_model()\n",
    "        \n",
    "        # Set objective in universal if told by user\n",
    "        # Made constraint as fraction of minimum in next step\n",
    "        if obj_constraint == True:\n",
    "            universal.add_reactions([model.reactions.get_by_id(obj)])\n",
    "            universal.objective = obj\n",
    "            orig_rxn_ids.remove(obj)\n",
    "            orig_rxns = []\n",
    "            for rxn in orig_rxn_ids: \n",
    "                orig_rxns.append(copy.deepcopy(model.reactions.get_by_id(rxn)))\n",
    "        else:\n",
    "            orig_rxns = list(copy.deepcopy(model.reactions))\n",
    "            \n",
    "        # Add pFBA to universal model and add model reactions\n",
    "        print('Add pFBA...')\n",
    "        add_pfba_likely(universal, likelihoods)\n",
    "        \n",
    "#         updated_universal = copy.deepcopy(universal)\n",
    "#         universal = copy.deepcopy(universal) # reset solver\n",
    "        universal.add_reactions(orig_rxns)\n",
    "        \n",
    "        # If previous objective not set as constraint, set minimum lower bound\n",
    "        if obj_constraint == False: \n",
    "            universal.reactions.get_by_id(obj).lower_bound = obj_lb\n",
    "    \n",
    "        # Set metabolic tasks that must carry flux in gapfilled solution\n",
    "        if tasks != None:\n",
    "            for task in tasks:                    \n",
    "                universal.reactions.get_by_id(task).lower_bound = task_lb\n",
    "                \n",
    "        # Run FBA and save solution\n",
    "        print('Optimizing model with combined reactions...')\n",
    "        solution = universal.optimize()\n",
    "\n",
    "        if iters > 1:\n",
    "            print('Generating flux sampling object...')\n",
    "            optgp_object = OptGPSampler(universal, processes=4)\n",
    "        \n",
    "            # Assess the sampled flux distributions\n",
    "            print('Sampling ' + str(iters) + ' flux distributions...')\n",
    "            flux_samples = optgp_object.sample(iters)\n",
    "            rxns = list(flux_samples.columns)\n",
    "            for distribution in flux_samples.iterrows():\n",
    "                for flux in range(0, len(list(distribution[1]))):\n",
    "                    if abs(list(distribution[1])[flux]) > 1e-6:\n",
    "                        new_rxn_ids |= set([rxns[flux]]).difference(orig_rxn_ids)\n",
    "        else:\n",
    "            rxns = list(solution.fluxes.index)\n",
    "            fluxes = list(solution.fluxes)\n",
    "            for flux in range(0, len(fluxes)):\n",
    "                if abs(fluxes[flux]) > 1e-6:\n",
    "                    new_rxn_ids |= set([rxns[flux]])\n",
    "    \n",
    "    # Screen new reaction IDs\n",
    "    if obj in new_rxn_ids: new_rxn_ids.remove(obj)\n",
    "    for rxn in orig_rxn_ids:\n",
    "        try:\n",
    "            new_rxn_ids.remove(rxn)\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    # Get reactions and metabolites to be added to the model\n",
    "    print('Retrieving reactions and metabolites needed for gapfilling...')\n",
    "    new_rxns = copy.deepcopy([reaction_bag.reactions.get_by_id(rxn) for rxn in new_rxn_ids])\n",
    "    new_cpd_ids = set()\n",
    "    for rxn in new_rxns: new_cpd_ids |= set([str(x.id) for x in list(rxn.metabolites)])\n",
    "    new_cpd_ids = new_cpd_ids.difference(orig_cpd_ids)\n",
    "    new_cpds = copy.deepcopy([reaction_bag.metabolites.get_by_id(cpd) for cpd in new_cpd_ids])\n",
    "    \n",
    "    # Copy model and gapfill \n",
    "    print('Gapfilling model...')\n",
    "    new_model = copy.deepcopy(model)\n",
    "    new_model.add_metabolites(new_cpds)\n",
    "    new_model.add_reactions(new_rxns)\n",
    "    \n",
    "    # Identify extracellular metabolites with no exchanges\n",
    "    if add_exchanges == True:\n",
    "        new_exchanges = extend_exchanges(new_model, new_cpd_ids, extracellular)\n",
    "        if len(new_exchanges) > 0: new_rxn_ids |= new_exchanges\n",
    "    \n",
    "    duration = int(round(time.time() - start_time))\n",
    "    print('Took ' + str(duration) + ' seconds to gapfill ' + str(len(new_rxn_ids)) + \\\n",
    "          ' reactions and ' + str(len(new_cpd_ids)) + ' metabolites.') \n",
    "    \n",
    "    new_obj_val = new_model.slim_optimize()\n",
    "    if new_obj_val > 1e-6:\n",
    "        print('Gapfilled model objective now carries flux (' + str(new_obj_val) + ').')\n",
    "    else:\n",
    "        print('Gapfilled model objective still does not carry flux.')\n",
    "    \n",
    "    return {'NewModel':new_model, 'gaps':new_rxn_ids, 'mets':new_cpd_ids}\n",
    "\n",
    "\n",
    "# Adds missing exchanges for extracellulart metbaolites\n",
    "def extend_exchanges(model, cpd_ids, ex):\n",
    "    \n",
    "    model_exchanges = set(find_boundary_types(model, 'exchange', external_compartment=ex))\n",
    "    new_ex_ids = set()\n",
    "    \n",
    "    for cpd in cpd_ids:\n",
    "        cpd = model.metabolites.get_by_id(cpd)\n",
    "        if str(cpd.compartment) != ex:\n",
    "            continue\n",
    "        else:\n",
    "            if bool(set(cpd.reactions) & model_exchanges) == False:\n",
    "                try:\n",
    "                    new_id = 'EX_' + cpd.id\n",
    "                    model.add_boundary(cpd, type='exchange', reaction_id=new_id, lb=-1000.0, ub=1000.0)\n",
    "                    new_ex_ids |= set([new_id])\n",
    "                except ValueError:\n",
    "                    pass\n",
    "\n",
    "    return new_ex_ids\n",
    "\n",
    "\n",
    "# Returns the reaction ID of the objective reaction\n",
    "def get_objective(model):\n",
    "    \n",
    "    if len(list(model.objective.variables)) == 0:\n",
    "        raise IndexError('Model has no objective set.')\n",
    "    \n",
    "    expression = str(model.objective.expression).split()\n",
    "    if 'reverse' in expression[0]:\n",
    "        obj_id = expression[2].split('*')[-1]\n",
    "    else:\n",
    "        obj_id = expression[0].split('*')[-1]\n",
    "            \n",
    "    return obj_id\n",
    "\n",
    "from __future__ import absolute_import\n",
    "\n",
    "import logging\n",
    "from warnings import warn\n",
    "from itertools import chain\n",
    "\n",
    "from optlang.symbolics import Zero\n",
    "\n",
    "from cobra.util import solver as sutil\n",
    "from cobra.core.solution import get_solution\n",
    "\n",
    "LOGGER = logging.getLogger(__name__)\n",
    "\n",
    "def add_pfba_likely(model, likelihoods, objective=None, fraction_of_optimum=1.0):\n",
    "    \"\"\"Add pFBA objective\n",
    "\n",
    "    Add objective to minimize the summed flux of all reactions to the\n",
    "    current objective.\n",
    "\n",
    "    See Also\n",
    "    -------\n",
    "    pfba\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    model : cobra.Model\n",
    "        The model to add the objective to\n",
    "    objective :\n",
    "        An objective to set in combination with the pFBA objective.\n",
    "    fraction_of_optimum : float\n",
    "        Fraction of optimum which must be maintained. The original objective\n",
    "        reaction is constrained to be greater than maximal_value *\n",
    "        fraction_of_optimum.\n",
    "    \"\"\"\n",
    "    if objective is not None:\n",
    "        model.objective = objective\n",
    "    if model.solver.objective.name == '_pfba_objective':\n",
    "        raise ValueError('The model already has a pFBA objective.')\n",
    "    sutil.fix_objective_as_constraint(model, fraction=fraction_of_optimum)\n",
    "    reaction_variables = ((rxn.forward_variable, rxn.reverse_variable)\n",
    "                          for rxn in model.reactions)\n",
    "    variables = chain(*reaction_variables)\n",
    "    dict1 = {}\n",
    "    fail_report = []\n",
    "    model_reactions = [rxn.id.split('_')[0] for rxn in model.reactions if rxn.id.startswith('rxn')]\n",
    "    for v in variables:\n",
    "#         print(set([str(v.name.split('_')[0])]))\n",
    "#         print(model_reactions[0])\n",
    "        if set([str(v.name.split('_')[0])]).issubset(set(model_reactions)) and str(v.name.split('_')[0]).startswith('rxn'):\n",
    "            rxn_id = (v.name.split('_')[0] + '_c')\n",
    "#             print(rxn_id)\n",
    "#         for rxn in model.reactions:\n",
    "#             if v.name.startswith(rxn.id):\n",
    "            try:\n",
    "                dict1[v] = max(0.0, 1.0 - likelihoods[rxn_id])\n",
    "#                 print('likelihood added')\n",
    "            except:\n",
    "                try:\n",
    "                    dict1[v] = 1.0\n",
    "                except:\n",
    "                    print('FAILED')\n",
    "                    pass\n",
    "                pass\n",
    "        elif str(v.name.split('_')[0]).startswith('DM'):\n",
    "            dict1[v] = 1.0\n",
    "        else:\n",
    "            fail_report.append(1)\n",
    "\n",
    "    model.objective = model.problem.Objective(\n",
    "        Zero, direction='min', sloppy=True, name=\"_pfba_objective\")\n",
    "    model.objective.set_linear_coefficients(dict1)\n",
    "#     return dict1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genome_id = '220668.9'\n",
    "# model = cobra.io.read_sbml_model('../gap_models/'+ genome_id +'.xml')\n",
    "# likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "\n",
    "# dict_out = add_pfba_likely(model, likelihoods)\n",
    "\n",
    "# dict_out\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading in models...Adding Water...Set-up Universal...Set-up Model...0.535631914934mins to complete\n"
     ]
    }
   ],
   "source": [
    "# Using pFBA with new media components (RNA bases + thymidine...) \n",
    "# Remove reaction likelihoods of zero from model\n",
    "# Add demands for all metabolites in model to avoid any reactions being blocked\n",
    "# Use thresholded likelihoods to penalize reactions with no likelihood using pFBA gapfill\n",
    "# Use pFBA answer + 0 likelihood reactions from reconstructed model + higher likelihood reactions to probanno gapfill\n",
    "\n",
    "# Weighted pFBA gapfill solution\n",
    "\n",
    "t = time.time()\n",
    "# counter = 0\n",
    "\n",
    "sys.stdout.write('Loading in models...')\n",
    "\n",
    "universal = cobra.io.load_json_model(\"../Data/GramPosUni.json\")\n",
    "genome_id = '220668.9'\n",
    "model = cobra.io.read_sbml_model('../gap_models/'+ genome_id +'.xml')\n",
    "likelihoods = pickle.load(open('../likelihoods/'+ genome_id +'.probs'))\n",
    "\n",
    "sys.stdout.write('Adding Water...')\n",
    "\n",
    "# Ensure free diffusion of water\n",
    "model.reactions.get_by_id('rxn05319_c').name = \"Water transport\"\n",
    "model.reactions.get_by_id('rxn05319_c').bounds = (-1000., 1000.)\n",
    "\n",
    "sys.stdout.write('Set-up Universal...')\n",
    "\n",
    "# Add demand for all metabolites in Universal model to stop blocked reactions\n",
    "all_mets = []\n",
    "for met in universal.metabolites:\n",
    "    if (met.id.endswith('_c')):\n",
    "        universal.add_boundary(met, type='demand')\n",
    "\n",
    "### Set Up Model: remove low likelihood reactions\n",
    "sys.stdout.write('Set-up Model...')\n",
    "low_like_model = []\n",
    "for rxn in model.reactions:\n",
    "    if rxn.id.startswith('rxn'):\n",
    "        try:\n",
    "            if likelihoods[rxn.id] <= 0.1:\n",
    "                low_like_model.append(rxn.id)\n",
    "        except:\n",
    "            pass\n",
    "model_rxns_to_remove = [model.reactions.get_by_id(rxn) for rxn in low_like_model]\n",
    "model.remove_reactions(model_rxns_to_remove)\n",
    "\n",
    "# metabolite = model.metabolites.get_by_id('cpd00065_c')\n",
    "# demand = model.add_boundary(metabolite, type='demand')\n",
    "# model.objective = demand\n",
    "\n",
    "# Create specific Media List\n",
    "media_list = bsm + M9_sources + rna_bases # + nitrogen + carbon\n",
    "set_media(model, media_list, universal, verbose=False)\n",
    "set_media(universal, media_list, universal, verbose=False)\n",
    "\n",
    "print(str((time.time() - t)/60) + 'mins to complete')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>GramPositive.modeltemplate</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f4d0d37fd50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>6884</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>15041</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, e</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model GramPositive.modeltemplate at 0x7f4d0d37fd50>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# universal = cobra.io.load_json_model(\"../Data/GramPosUni.json\")\n",
    "# [rxn.id for rxn in universal.reactions if not str(rxn.id).startswith('rxn')]\n",
    "universal\n",
    "# 6884 mets; 8657 rxns;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# global_time = time.time()\n",
    "\n",
    "# new_gapfill_data = pfba_gapfill(model, universal, likelihoods, obj=None, obj_lb=10., obj_constraint=False, iters=1, add_exchanges=False)\n",
    "\n",
    "# print(str((time.time() - global_time)/60) + 'mins to complete')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting Loop...\n",
      "Loop0 Gapfilling...Creating universal model...\n",
      "Add pFBA...\n",
      "Optimizing model with combined reactions...\n",
      "Retrieving reactions and metabolites needed for gapfilling...\n",
      "Gapfilling model...\n",
      "Took 55 seconds to gapfill 2 reactions and 0 metabolites.\n",
      "Gapfilled model objective now carries flux (1000.0).\n",
      "pFBA...0.0...Constructing Dict...Ave likelihood of: cpd00117_c is 0.5173929430770764Run time: 1.46228676637 [mins]\n",
      "Loop1 Gapfilling...Creating universal model...\n",
      "Add pFBA...\n"
     ]
    }
   ],
   "source": [
    "# Run through each amino acid to check for production\n",
    "global_time = time.time()\n",
    "aa_like = {}\n",
    "counter = 0\n",
    "sys.stdout.write('Starting Loop...')\n",
    "\n",
    "total_dataset_dict = {}\n",
    "carb_idx = 0\n",
    "nit_idx = 0\n",
    "product_idx = 0\n",
    "carbon = 'D-Glucose'\n",
    "nitrogen = 'NH3'\n",
    "\n",
    "for aa_list in aas:\n",
    "    \n",
    "    sys.stdout.write('\\n'+ 'Loop' + str(counter) + ' ')\n",
    "    aa = aa_list[1]+'_c'\n",
    "    product = aa_list[0]\n",
    "    \n",
    "    with model as temp_model:\n",
    "        metabolite = temp_model.metabolites.get_by_id(aa)\n",
    "        demand = temp_model.add_boundary(metabolite, type='demand')\n",
    "        temp_model.objective = demand\n",
    "\n",
    "        sys.stdout.write('Gapfilling...')\n",
    "        dont_continue = 0\n",
    "        new_gapfill_data = pfba_gapfill(temp_model, universal, likelihoods, obj=None, obj_lb=10., obj_constraint=False, iters=1, add_exchanges=False)\n",
    "\n",
    "#         try:\n",
    "#             new_gapfill_data = pfba_gapfill(temp_model, universal, likelihoods, obj=None, obj_lb=10., obj_constraint=False, iters=1, add_exchanges=False)\n",
    "#         except:\n",
    "#             try:\n",
    "#                 new_gapfill_data = pfba_gapfill(temp_model, universal, likelihoods, obj=None, obj_lb=10., obj_constraint=False, iters=1, add_exchanges=False)\n",
    "#             except:\n",
    "#                 dont_continue = 1\n",
    "#                 pass\n",
    "\n",
    "        gaps_to_fill = new_gapfill_data['gaps']\n",
    "        new_model = new_gapfill_data['NewModel']\n",
    "        mets_added = new_gapfill_data['mets']\n",
    "\n",
    "        if dont_continue == 0:\n",
    "            # Optimize with filled pathway\n",
    "            sys.stdout.write('pFBA...')\n",
    "            solution = pfba(new_model, objective = demand)\n",
    "            sys.stdout.write(str(round(model.slim_optimize())) + '...')\n",
    "\n",
    "            sys.stdout.write('Constructing Dict...')\n",
    "            # Find reactions that carry flux\n",
    "            df = solution.fluxes.to_frame()\n",
    "            active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "            \n",
    "            demand_list = []\n",
    "            for rxn_id in active.index:\n",
    "                if rxn_id.startswith('DM') and rxn_id != demand.id:\n",
    "                    demand_list.append(rxn_id)\n",
    "                    print('demands added')\n",
    "\n",
    "            # Acquire likelihood scores for reactions that carry flux\n",
    "            flux_rxns = []\n",
    "            like_list = []\n",
    "            gap_flux_rxns = []\n",
    "            gap_like_list = []\n",
    "            path_flux_rxns = []\n",
    "            path_like_list = []\n",
    "            for rxn in list(active.index):\n",
    "                if rxn in gaps_to_fill and rxn.startswith('rxn'):\n",
    "                    try:\n",
    "                        gap_flux_rxns.append([str(rxn),likelihoods[str(rxn)]])\n",
    "                        gap_like_list.append(likelihoods[str(rxn)])\n",
    "#                         print('success')\n",
    "                    except:\n",
    "                        pass\n",
    "                if rxn not in gaps_to_fill and rxn.startswith('rxn'):\n",
    "                    try:\n",
    "                        path_flux_rxns.append([str(rxn),likelihoods[str(rxn)]])\n",
    "                        path_like_list.append(likelihoods[str(rxn)])\n",
    "#                         print('success2')\n",
    "                    except:\n",
    "                        pass\n",
    "                if rxn.startswith('rxn'):\n",
    "                    try:\n",
    "                        flux_rxns.append([str(rxn),likelihoods[str(rxn)]])\n",
    "                        like_list.append(likelihoods[str(rxn)])\n",
    "                    except:\n",
    "                        pass\n",
    "            avg_like = np.mean(like_list)\n",
    "            gap_avg_like = np.mean(gap_like_list)\n",
    "            path_avg_like = np.mean(path_like_list)\n",
    "            sys.stdout.write('Ave likelihood of: ' + aa + ' is ' + str(avg_like))\n",
    "\n",
    "            counter += 1\n",
    "\n",
    "            report_dict = {}\n",
    "\n",
    "            report_dict['Model_ID'] = genome_id\n",
    "            report_dict['Carbon'] = carbon\n",
    "            report_dict['Nitrogen'] = nitrogen\n",
    "            report_dict['objective'] = product\n",
    "            report_dict['opt_before'] = model.slim_optimize()\n",
    "            report_dict['opt_after'] = new_model.slim_optimize()\n",
    "            report_dict['avg_path_like'] = avg_like\n",
    "            report_dict['gap_avg_like'] = gap_avg_like\n",
    "            report_dict['path_avg_like'] = path_avg_like\n",
    "            report_dict['gaps_filled'] = gaps_to_fill\n",
    "            report_dict['mets_added'] = mets_added\n",
    "            report_dict['reactions_w_flux'] = flux_rxns\n",
    "            report_dict['gaps_w_flux'] = gap_flux_rxns\n",
    "            report_dict['path_w_flux'] = path_flux_rxns\n",
    "            report_dict['active_rxns'] = active\n",
    "            report_dict['demands'] = demand_list\n",
    "\n",
    "            report_dict_ID = genome_id + ':' + str(carb_idx) + '.' + str(nit_idx) + '.' + str(product_idx)\n",
    "            total_dataset_dict[report_dict_ID] = report_dict\n",
    "            product_idx += 1 #Keep track to which product is being maximized\n",
    "\n",
    "            elapsed = time.time() - t\n",
    "            sys.stdout.write('Run time: ' + str(elapsed/60) + \" [mins]\")\n",
    "\n",
    "        elif dont_continue == 1:\n",
    "            sys.stdout.write('Failed to gapfill...')\n",
    "            counter += 1\n",
    "\n",
    "            report_dict = {}\n",
    "\n",
    "            report_dict['Model_ID'] = genome_id\n",
    "            report_dict['Carbon'] = carbon\n",
    "            report_dict['Nitrogen'] = nitrogen\n",
    "            report_dict['objective'] = product\n",
    "            report_dict['opt_before'] = \"Failed to gapfill\"\n",
    "            report_dict['opt_after'] = \"Failed to gapfill\"\n",
    "            report_dict['avg_path_like'] = \"Failed to gapfill\"\n",
    "            report_dict['gap_avg_like'] = \"Failed to gapfill\"\n",
    "            report_dict['path_avg_like'] = \"Failed to gapfill\"\n",
    "            report_dict['gaps_filled'] = \"Failed to gapfill\"\n",
    "            report_dict['mets_added'] = \"Failed to gapfill\"\n",
    "            report_dict['reactions_w_flux'] = \"Failed to gapfill\"\n",
    "            report_dict['gaps_w_flux'] = \"Failed to gapfill\"\n",
    "            report_dict['path_w_flux'] = \"Failed to gapfill\"\n",
    "            report_dict['active_rxns'] = \"Failed to gapfill\"\n",
    "            report_dict['demands'] = \"Failed to gapfill\"\n",
    "\n",
    "            report_dict_ID = genome_id + ':' + str(carb_idx) + '.' + str(nit_idx) + '.' + str(product_idx)\n",
    "            total_dataset_dict[report_dict_ID] = report_dict\n",
    "            product_idx += 1 #Keep track to which product is being maximized\n",
    "\n",
    "            elapsed = time.time() - t\n",
    "            sys.stdout.write('Run time: ' + str(elapsed/60) + \" [mins]\")\n",
    "\n",
    "file_name = \"../metabolic_output/%s.data\" % (genome_id)\n",
    "pickle.dump(total_dataset_dict, open(file_name, \"wb\"))\n",
    "\n",
    "# elapsed = time.time() - t\n",
    "# print(\"\\nTime to complete: \" + str(elapsed/60) + \" [mins]\")\n",
    "print(str((time.time() - global_time)/60) + 'mins to complete')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_dataset_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hours = 34*9*49/60\n",
    "hours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#     ['D-Alanine','cpd00117'], #EXs : \n",
    "#     ['D-Glutamate','cpd00186'], #EXs : \n",
    "#     ['D-Methionine','cpd00637'], #EXs : \n",
    "#     ['D-Serine','cpd00550'], #EXs : \n",
    "#     ['Glycine','cpd00033'], #EXs : 1\n",
    "#     ['L-Alanine','cpd00035'], #EXs : 2\n",
    "#     ['L-Arginine','cpd00051'], #EXs : 3\n",
    "#     ['L-Asparagine','cpd00132'], #EXs : 4\n",
    "#     ['L-Aspartate','cpd00041'], #EXs : 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(model.slim_optimize())\n",
    "# solution = model.optimize()\n",
    "# df = solution.fluxes.to_frame()\n",
    "# active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "\n",
    "# active\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # print(new_model.slim_optimize())\n",
    "# # solution = new_model.optimize()\n",
    "# solution = pfba(new_model, objective = demand)\n",
    "# df = solution.fluxes.to_frame()\n",
    "# active = df.loc[(abs(df['fluxes'])) > 0.1]\n",
    "# print(str(len(active.index)))\n",
    "# active\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for rxn in active.index:\n",
    "#     try:\n",
    "#         if rxn.startswith('rxn'):\n",
    "#             print(rxn)\n",
    "#             print(likelihoods[rxn])\n",
    "#     except:\n",
    "#         pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Acquire likelihood scores for reactions that carry flux\n",
    "# aa = ' the amino acid'\n",
    "# flux_rxns = []\n",
    "# like_list = []\n",
    "# for rxn in list(active.index):\n",
    "#     if rxn.startswith('rxn'):\n",
    "#         try:\n",
    "#             flux_rxns.append([str(rxn),likelihoods[str(rxn)]])\n",
    "#             like_list.append(likelihoods[str(rxn)])\n",
    "#         except:\n",
    "#             pass\n",
    "# avg_like = np.mean(like_list)\n",
    "# sys.stdout.write('Ave likelihood of: ' + aa + ' is ' + str(avg_like))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
